diff --git a/src/plugins/platforms/windows/uiautomation/qwindowsuiaselectionitemprovider.cpp b/src/plugins/platforms/windows/uiautomation/qwindowsuiaselectionitemprovider.cpp
index 30e9524d54e..71e417c90d6 100644
--- a/src/plugins/platforms/windows/uiautomation/qwindowsuiaselectionitemprovider.cpp
+++ b/src/plugins/platforms/windows/uiautomation/qwindowsuiaselectionitemprovider.cpp
@@ -180,10 +180,6 @@ HRESULT STDMETHODCALLTYPE QWindowsUiaSelectionItemProvider::get_SelectionContain
     if (!accessible)
         return UIA_E_ELEMENTNOTAVAILABLE;
 
-    QAccessibleActionInterface *actionInterface = accessible->actionInterface();
-    if (!actionInterface)
-        return UIA_E_ELEMENTNOTAVAILABLE;
-
     // Radio buttons do not require a container.
     if (accessible->role() == QAccessible::ListItem) {
         if (QAccessibleInterface *parent = accessible->parent()) {
