diff --git a/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp b/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp
--- a/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp
+++ b/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp
@@ -203,9 +204,9 @@ void QWindowsUiaMainProvider::notifyValueChange(QAccessibleValueChangeEvent *eve
 void QWindowsUiaMainProvider::notifyNameChange(QAccessibleEvent *event)
 {
     if (QAccessibleInterface *accessible = event->accessibleInterface()) {
-        // Restrict notification to combo boxes, which need it for accessibility,
-        // in order to avoid slowdowns with unnecessary notifications.
-        if (accessible->role() == QAccessible::ComboBox) {
+        // Restrict notification to combo boxes and the currently focused element, which
+        // need it for accessibility, in order to avoid slowdowns with unnecessary notifications.
+        if (accessible->role() == QAccessible::ComboBox || accessible->state().focused) {
             if (QWindowsUiaMainProvider *provider = providerForAccessible(accessible)) {
                 VARIANT oldVal, newVal;
                 clearVariant(&oldVal);
