diff --git a/src/platformsupport/fontdatabases/windows/qwindowsfontdatabase.cpp b/src/platformsupport/fontdatabases/windows/qwindowsfontdatabase.cpp
index 09d2d916fe..195f232354 100644
--- a/src/platformsupport/fontdatabases/windows/qwindowsfontdatabase.cpp
+++ b/src/platformsupport/fontdatabases/windows/qwindowsfontdatabase.cpp
@@ -1710,51 +1710,75 @@ HFONT QWindowsFontDatabase::systemFont()
 
 static const char *other_tryFonts[] = {
     "Arial",
-    "MS UI Gothic",
-    "Gulim",
-    "SimSun",
-    "PMingLiU",
+    "Yu Gothic UI",
+    "Meiryo UI",
+    "Yu Gothic",
+    "Meiryo",
+    "Malgun Gothic",
+    "Microsoft YaHei UI",
+    "Microsoft YaHei",
+    "Microsoft JhengHei UI",
+    "Microsoft JhengHei",
     "Arial Unicode MS",
     0
 };
 
 static const char *jp_tryFonts [] = {
     "Yu Gothic UI",
-    "MS UI Gothic",
+    "Meiryo UI",
+    "Yu Gothic",
+    "Meiryo",
     "Arial",
-    "Gulim",
-    "SimSun",
-    "PMingLiU",
+    "Malgun Gothic",
+    "Microsoft YaHei UI",
+    "Microsoft YaHei",
+    "Microsoft JhengHei UI",
+    "Microsoft JhengHei",
     "Arial Unicode MS",
     0
 };
 
 static const char *ch_CN_tryFonts [] = {
-    "SimSun",
+    "Microsoft YaHei UI",
+    "Microsoft YaHei",
     "Arial",
-    "PMingLiU",
-    "Gulim",
-    "MS UI Gothic",
+    "Microsoft JhengHei UI",
+    "Microsoft JhengHei",
+    "Malgun Gothic",
+    "Yu Gothic UI",
+    "Meiryo UI",
+    "Yu Gothic",
+    "Meiryo",
     "Arial Unicode MS",
     0
 };
 
 static const char *ch_TW_tryFonts [] = {
-    "PMingLiU",
+    "Microsoft JhengHei UI",
+    "Microsoft JhengHei",
     "Arial",
-    "SimSun",
-    "Gulim",
-    "MS UI Gothic",
+    "Microsoft YaHei UI",
+    "Microsoft YaHei",
+    "Malgun Gothic",
+    "Yu Gothic UI",
+    "Meiryo UI",
+    "Yu Gothic",
+    "Meiryo",
     "Arial Unicode MS",
     0
 };
 
 static const char *kr_tryFonts[] = {
-    "Gulim",
+    "Malgun Gothic",
     "Arial",
-    "PMingLiU",
-    "SimSun",
-    "MS UI Gothic",
+    "Microsoft JhengHei UI",
+    "Microsoft JhengHei",
+    "Microsoft YaHei UI",
+    "Microsoft YaHei",
+    "Yu Gothic UI",
+    "Meiryo UI",
+    "Yu Gothic",
+    "Meiryo",
     "Arial Unicode MS",
     0
 };
@@ -1892,6 +1916,12 @@ QStringList QWindowsFontDatabase::extraTryFontsForFamily(const QString &family)
             ++tf;
         }
     }
+    LOGFONT lf;
+    GetObject(QWindowsFontDatabase::systemFont(), sizeof(lf), &lf);
+    result.prepend(QString::fromWCharArray(lf.lfFaceName));
+    result.append(QStringLiteral("Nirmala UI"));
+    result.append(QStringLiteral("Iskoola Pota"));
+    result.append(QStringLiteral("Ebrima"));
     result.append(QStringLiteral("Segoe UI Emoji"));
     result.append(QStringLiteral("Segoe UI Symbol"));
     return result;
