diff --git a/src/platformsupport/fontdatabases/windows/qwindowsfontdatabase.cpp b/src/platformsupport/fontdatabases/windows/qwindowsfontdatabase.cpp
index 9e6e5d88c7..e6c8519a23 100644
--- a/src/platformsupport/fontdatabases/windows/qwindowsfontdatabase.cpp
+++ b/src/platformsupport/fontdatabases/windows/qwindowsfontdatabase.cpp
@@ -1694,50 +1694,90 @@ HFONT QWindowsFontDatabase::systemFont()
 
 static const char *other_tryFonts[] = {
     "Arial",
-    "MS UI Gothic",
-    "Gulim",
-    "SimSun",
-    "PMingLiU",
+    "Yu Gothic UI",
+    "Meiryo UI",
+    "Yu Gothic",
+    "Meiryo",
+    "Malgun Gothic",
+    "Microsoft YaHei UI",
+    "Microsoft YaHei",
+    "Microsoft JhengHei UI",
+    "Microsoft JhengHei",
+    "Nirmala UI",
+    "Iskoola Pota",
+    "Ebrima",
     "Arial Unicode MS",
     0
 };
 
 static const char *jp_tryFonts [] = {
-    "MS UI Gothic",
+    "Yu Gothic UI",
+    "Meiryo UI",
+    "Yu Gothic",
+    "Meiryo",
     "Arial",
-    "Gulim",
-    "SimSun",
-    "PMingLiU",
+    "Malgun Gothic",
+    "Microsoft YaHei UI",
+    "Microsoft YaHei",
+    "Microsoft JhengHei UI",
+    "Microsoft JhengHei",
+    "Nirmala UI",
+    "Iskoola Pota",
+    "Ebrima",
     "Arial Unicode MS",
     0
 };
 
 static const char *ch_CN_tryFonts [] = {
-    "SimSun",
+    "Microsoft YaHei UI",
+    "Microsoft YaHei",
     "Arial",
-    "PMingLiU",
-    "Gulim",
-    "MS UI Gothic",
+    "Microsoft JhengHei UI",
+    "Microsoft JhengHei",
+    "Malgun Gothic",
+    "Yu Gothic UI",
+    "Meiryo UI",
+    "Yu Gothic",
+    "Meiryo",
+    "Nirmala UI",
+    "Iskoola Pota",
+    "Ebrima",
     "Arial Unicode MS",
     0
 };
 
 static const char *ch_TW_tryFonts [] = {
-    "PMingLiU",
+    "Microsoft JhengHei UI",
+    "Microsoft JhengHei",
     "Arial",
-    "SimSun",
-    "Gulim",
-    "MS UI Gothic",
+    "Microsoft YaHei UI",
+    "Microsoft YaHei",
+    "Malgun Gothic",
+    "Yu Gothic UI",
+    "Meiryo UI",
+    "Yu Gothic",
+    "Meiryo",
+    "Nirmala UI",
+    "Iskoola Pota",
+    "Ebrima",
     "Arial Unicode MS",
     0
 };
 
 static const char *kr_tryFonts[] = {
-    "Gulim",
+    "Malgun Gothic",
     "Arial",
-    "PMingLiU",
-    "SimSun",
-    "MS UI Gothic",
+    "Microsoft JhengHei UI",
+    "Microsoft JhengHei",
+    "Microsoft YaHei UI",
+    "Microsoft YaHei",
+    "Yu Gothic UI",
+    "Meiryo UI",
+    "Yu Gothic",
+    "Meiryo",
+    "Nirmala UI",
+    "Iskoola Pota",
+    "Ebrima",
     "Arial Unicode MS",
     0
 };
