diff --git a/src/corelib/text/qstringconverter.cpp b/src/corelib/text/qstringconverter.cpp
index 73fe55d07c..5121a09dcd 100644
--- a/src/corelib/text/qstringconverter.cpp
+++ b/src/corelib/text/qstringconverter.cpp
@@ -2123,7 +2123,7 @@ QStringDecoder QStringDecoder::decoderForHtml(QByteArrayView data)
     std::optional<QStringConverter::Encoding> encoding = encodingForData(data);
     if (encoding)
         // trust the initial BOM
-        return QStringDecoder(encoding.value());
+        return QStringDecoder(*encoding);
 
     QByteArray encodingTag = parseHtmlMetaForEncoding(data);
     if (!encodingTag.isEmpty())
diff --git a/src/xml/dom/qdom.cpp b/src/xml/dom/qdom.cpp
index 3c41ed48a3..e0e386cafb 100644
--- a/src/xml/dom/qdom.cpp
+++ b/src/xml/dom/qdom.cpp
@@ -5844,7 +5844,7 @@ void QDomDocumentPrivate::saveDocument(QTextStream& s, const int indent, QDomNod
                 if (!encoding)
                     qWarning() << "QDomDocument::save(): Unsupported encoding" << enc << "specified.";
                 else
-                    s.setEncoding(encoding.value());
+                    s.setEncoding(*encoding);
             }
         }
 #endif
