diff --git a/src/gui/platform/unix/dbustray/qstatusnotifieritemadaptor.cpp b/src/gui/platform/unix/dbustray/qstatusnotifieritemadaptor.cpp
index 27a14e408a..2f6c13b6cf 100644
--- a/src/gui/platform/unix/dbustray/qstatusnotifieritemadaptor.cpp
+++ b/src/gui/platform/unix/dbustray/qstatusnotifieritemadaptor.cpp
@@ -136,6 +136,12 @@ void QStatusNotifierItemAdaptor::ContextMenu(int x, int y)
     emit m_trayIcon->activated(QPlatformSystemTrayIcon::Context);
 }
 
+void QStatusNotifierItemAdaptor::ProvideXdgActivationToken(const QString &token)
+{
+    qCDebug(qLcTray) << token;
+    qputenv("XDG_ACTIVATION_TOKEN", token.toUtf8());
+}
+
 void QStatusNotifierItemAdaptor::Scroll(int w, const QString &s)
 {
     qCDebug(qLcTray) << w << s;
diff --git a/src/gui/platform/unix/dbustray/qstatusnotifieritemadaptor_p.h b/src/gui/platform/unix/dbustray/qstatusnotifieritemadaptor_p.h
index 286aafa9bb..103fc974dd 100644
--- a/src/gui/platform/unix/dbustray/qstatusnotifieritemadaptor_p.h
+++ b/src/gui/platform/unix/dbustray/qstatusnotifieritemadaptor_p.h
@@ -73,6 +73,9 @@ class QStatusNotifierItemAdaptor: public QDBusAbstractAdaptor
 "    <property access=\"read\" type=\"(sa(iiay)ss)\" name=\"ToolTip\">\n"
 "      <annotation value=\"QXdgDBusToolTipStruct\" name=\"org.qtproject.QtDBus.QtTypeName\"/>\n"
 "    </property>\n"
+"    <method name=\"ProvideXdgActivationToken\">\n"
+"        <arg name=\"token\" type=\"s\" direction=\"in\"/>\n"
+"    </method>\n"
 "    <method name=\"ContextMenu\">\n"
 "      <arg direction=\"in\" type=\"i\" name=\"x\"/>\n"
 "      <arg direction=\"in\" type=\"i\" name=\"y\"/>\n"
@@ -150,6 +153,7 @@ public: // PROPERTIES
 public Q_SLOTS: // METHODS
     void Activate(int x, int y);
     void ContextMenu(int x, int y);
+    void ProvideXdgActivationToken(const QString &token);
     void Scroll(int delta, const QString &orientation);
     void SecondaryActivate(int x, int y);
 Q_SIGNALS: // SIGNALS
