diff --git a/src/widgets/kernel/qwidget.cpp b/src/widgets/kernel/qwidget.cpp
index 202750d659..5994104926 100644
--- a/src/widgets/kernel/qwidget.cpp
+++ b/src/widgets/kernel/qwidget.cpp
@@ -1085,8 +1085,8 @@ static bool q_evaluateRhiConfigRecursive(const QWidget *w, QPlatformBackingStore
     }
     QObjectList children = w->children();
     for (int i = 0; i < children.size(); i++) {
-        if (children.at(i)->isWidgetType()) {
-            const QWidget *childWidget = qobject_cast<const QWidget *>(children.at(i));
+        if (const QObject *child = children.at(i); child && child->isWidgetType()) {
+            const QWidget *childWidget = qobject_cast<const QWidget *>(child);
             if (childWidget) {
                 if (q_evaluateRhiConfigRecursive(childWidget, outConfig, outType))
                     return true;
