diff --git a/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportalfiledialog.cpp b/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportalfiledialog.cpp
index 54e600fcbe..9e3afea7c5 100644
--- a/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportalfiledialog.cpp
+++ b/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportalfiledialog.cpp
@@ -173,9 +173,12 @@ void QXdgDesktopPortalFileDialog::openPortal(Qt::WindowFlags windowFlags, Qt::Wi
 
         if (!d->selectedFiles.isEmpty()) {
             // current_file for the file to be pre-selected, current_name for the file name to be pre-filled
-            // current_file accepts absolute path while current_name accepts just file name
-            options.insert("current_file"_L1, QFile::encodeName(d->selectedFiles.first()).append('\0'));
-            options.insert("current_name"_L1, QFileInfo(d->selectedFiles.first()).fileName());
+            // current_file accepts absolute path and requires the file to exist
+            // while current_name accepts just file name
+            QFileInfo selectedFileInfo(d->selectedFiles.first());
+            if (selectedFileInfo.exists())
+                options.insert("current_file"_L1, QFile::encodeName(d->selectedFiles.first()).append('\0'));
+            options.insert("current_name"_L1, selectedFileInfo.fileName());
         }
     }
 
