diff --git a/src/gui/CMakeLists.txt b/src/gui/CMakeLists.txt
index 1ce10a3e850..de4b2715ba1 100644
--- a/src/gui/CMakeLists.txt
+++ b/src/gui/CMakeLists.txt
@@ -217,6 +217,7 @@ qt_internal_add_module(Gui
         painting/qpixellayout.cpp painting/qpixellayout_p.h
         painting/qplatformbackingstore.cpp painting/qplatformbackingstore.h
         painting/qpolygon.cpp painting/qpolygon.h
+        painting/qrasterbackingstore.cpp painting/qrasterbackingstore_p.h
         painting/qrasterdefs_p.h
         painting/qrasterizer.cpp painting/qrasterizer_p.h
         painting/qrbtree_p.h
@@ -224,6 +225,7 @@ qt_internal_add_module(Gui
         painting/qrgb.h
         painting/qrgba64.h painting/qrgba64_p.h
         painting/qrgbafloat.h
+        painting/qrhibackingstore.cpp painting/qrhibackingstore_p.h
         painting/qstroker.cpp painting/qstroker_p.h
         painting/qtextureglyphcache.cpp painting/qtextureglyphcache_p.h
         painting/qtransform.cpp painting/qtransform.h
@@ -419,8 +421,6 @@ qt_internal_extend_target(Gui CONDITION APPLE
     SOURCES
         image/qimage_darwin.mm
         painting/qcoregraphics.mm painting/qcoregraphics_p.h
-        painting/qrasterbackingstore.cpp painting/qrasterbackingstore_p.h
-        painting/qrhibackingstore.cpp painting/qrhibackingstore_p.h
         platform/darwin/qmacmimeregistry.mm platform/darwin/qmacmimeregistry_p.h
         platform/darwin/qutimimeconverter.mm platform/darwin/qutimimeconverter.h
         platform/darwin/qapplekeymapper.mm platform/darwin/qapplekeymapper_p.h
diff --git a/src/gui/painting/qbackingstoredefaultcompositor.cpp b/src/gui/painting/qbackingstoredefaultcompositor.cpp
index fae9dabca66..9a335495f5f 100644
--- a/src/gui/painting/qbackingstoredefaultcompositor.cpp
+++ b/src/gui/painting/qbackingstoredefaultcompositor.cpp
@@ -99,8 +99,6 @@ QRhiTexture *QBackingStoreDefaultCompositor::toTexture(const QImage &sourceImage
 
     if (needsConversion)
         image = image.convertToFormat(QImage::Format_RGBA8888);
-    else
-        image.detach(); // if it was just wrapping data, that's no good, we need ownership, so detach
 
     if (resized) {
         if (!m_texture)
diff --git a/src/plugins/platforms/cocoa/qcocoaintegration.mm b/src/plugins/platforms/cocoa/qcocoaintegration.mm
index 350fe09e059..3fb0f013c0c 100644
--- a/src/plugins/platforms/cocoa/qcocoaintegration.mm
+++ b/src/plugins/platforms/cocoa/qcocoaintegration.mm
@@ -312,15 +312,6 @@ QPlatformBackingStore *QCocoaIntegration::createPlatformBackingStore(QWindow *wi
     case QSurface::MetalSurface:
     case QSurface::OpenGLSurface:
     case QSurface::VulkanSurface:
-        // If the window is a widget window, we know that the QWidgetRepaintManager
-        // will explicitly use rhiFlush() for the window owning the backingstore,
-        // and any child window with the same surface format. This means we can
-        // safely return a QCALayerBackingStore here, to ensure that any plain
-        // flush() for child windows that don't have a matching surface format
-        // will still work, by setting the layer's contents property.
-        if (window->inherits("QWidgetWindow"))
-            return new QCALayerBackingStore(window);
-
         // Otherwise we return a QRhiBackingStore, that implements flush() in
         // terms of rhiFlush().
         return new QRhiBackingStore(window);
diff --git a/src/plugins/platforms/wayland/qwaylandintegration.cpp b/src/plugins/platforms/wayland/qwaylandintegration.cpp
index 669d47ee3b8..01390d40d76 100644
--- a/src/plugins/platforms/wayland/qwaylandintegration.cpp
+++ b/src/plugins/platforms/wayland/qwaylandintegration.cpp
@@ -29,6 +29,7 @@
 #include <QtGui/private/qgenericunixtheme_p.h>
 
 #include <QtGui/private/qguiapplication_p.h>
+#include <QtGui/private/qrhibackingstore_p.h>
 
 #include <qpa/qwindowsysteminterface.h>
 #include <qpa/qplatformcursor.h>
@@ -166,6 +167,13 @@ QOpenGLContext *QWaylandIntegration::createOpenGLContext(EGLContext context, EGL
 
 QPlatformBackingStore *QWaylandIntegration::createPlatformBackingStore(QWindow *window) const
 {
+    switch (window->surfaceType()) {
+    case QSurface::OpenGLSurface:
+    case QSurface::VulkanSurface:
+        return new QRhiBackingStore(window);
+    default:
+        break;
+    }
     return new QWaylandShmBackingStore(window, mDisplay.data());
 }
 
diff --git a/src/plugins/platforms/windows/qwindowsgdiintegration.cpp b/src/plugins/platforms/windows/qwindowsgdiintegration.cpp
index 14fd03fc1f9..6c3b1f9342d 100644
--- a/src/plugins/platforms/windows/qwindowsgdiintegration.cpp
+++ b/src/plugins/platforms/windows/qwindowsgdiintegration.cpp
@@ -8,6 +8,7 @@
 
 #include <QtCore/qdebug.h>
 #include <QtGui/private/qpixmap_raster_p.h>
+#include <QtGui/private/qrhibackingstore_p.h>
 
 QT_BEGIN_NAMESPACE
 
@@ -37,6 +38,14 @@ QPlatformPixmap *QWindowsGdiIntegration::createPlatformPixmap(QPlatformPixmap::P
 
 QPlatformBackingStore *QWindowsGdiIntegration::createPlatformBackingStore(QWindow *window) const
 {
+    switch (window->surfaceType()) {
+    case QSurface::Direct3DSurface:
+    case QSurface::OpenGLSurface:
+    case QSurface::VulkanSurface:
+        return new QRhiBackingStore(window);
+    default:
+        break;
+    }
     return new QWindowsBackingStore(window);
 }
 
diff --git a/src/plugins/platforms/xcb/qxcbintegration.cpp b/src/plugins/platforms/xcb/qxcbintegration.cpp
index 67242c0c001..5334c3a3ac8 100644
--- a/src/plugins/platforms/xcb/qxcbintegration.cpp
+++ b/src/plugins/platforms/xcb/qxcbintegration.cpp
@@ -31,6 +31,7 @@
 #include <stdio.h>
 
 #include <QtGui/private/qguiapplication_p.h>
+#include <QtGui/private/qrhibackingstore_p.h>
 
 #if QT_CONFIG(xcb_xlib)
 #define register        /* C++17 deprecated register */
@@ -286,6 +287,13 @@ QPlatformBackingStore *QXcbIntegration::createPlatformBackingStore(QWindow *wind
         backingStore = new QXcbNativeBackingStore(window);
 #endif
     } else {
+        switch (window->surfaceType()) {
+        case QSurface::OpenGLSurface:
+        case QSurface::VulkanSurface:
+            backingStore = new QRhiBackingStore(window);
+        default:
+            break;
+        }
         backingStore = new QXcbBackingStore(window);
     }
     Q_ASSERT(backingStore);
