diff --git a/configure.cmake b/configure.cmake
index 5705e37611..ffc1a0e5c4 100644
--- a/configure.cmake
+++ b/configure.cmake
@@ -588,8 +588,7 @@ qt_feature("c11" PUBLIC
 )
 qt_feature("precompile_header"
     LABEL "Using precompiled headers"
-    CONDITION BUILD_WITH_PCH AND TEST_precompile_header
-    AUTODETECT NOT WASM
+    AUTODETECT BUILD_WITH_PCH AND TEST_precompile_header AND NOT WASM
 )
 qt_feature_config("precompile_header" QMAKE_PRIVATE_CONFIG)
 set(__qt_ltcg_detected FALSE)
