diff --git a/src/plugins/platforms/cocoa/qcocoawindow.mm b/src/plugins/platforms/cocoa/qcocoawindow.mm
index 9bf1ee9711..4b2c06fff9 100644
--- a/src/plugins/platforms/cocoa/qcocoawindow.mm
+++ b/src/plugins/platforms/cocoa/qcocoawindow.mm
@@ -536,12 +536,15 @@ NSUInteger QCocoaWindow::windowStyleMask(Qt::WindowFlags flags)
     }
 
     if (frameless) {
+        // Patch: Fix translucent Qt::ToolTip / Qt::Popup and possibly even Qt::Dialog.
+#if 0
         // Frameless windows do not display the traffic lights buttons for
         // e.g. minimize, however StyleMaskMiniaturizable is required to allow
         // programatic minimize. However, for framless tool windows (e.g. dock windows)
         // we don't want that, as it breaks translucency.
         if (type != Qt::Tool)
             styleMask |= NSWindowStyleMaskMiniaturizable;
+#endif
     } else if (flags & Qt::CustomizeWindowHint) {
         if (flags & Qt::WindowTitleHint)
             styleMask |= NSWindowStyleMaskTitled;
