diff --git a/src/gui/painting/qrasterizer.cpp b/src/gui/painting/qrasterizer.cpp
index 7b26b20db8..f992587e9a 100644
--- a/src/gui/painting/qrasterizer.cpp
+++ b/src/gui/painting/qrasterizer.cpp
@@ -795,6 +795,16 @@ void QRasterizer::rasterizeLine(const QPointF &a, const QPointF &b, qreal width,
 
         pa = npa;
         pb = npb;
+
+        // Patch: Workaround weird -O3 AppleClang bug, see t.me/ProCxx/475907
+        volatile auto log = false;
+        if (log) {
+            [[maybe_unused]] const auto touched = std::string()
+                + std::to_string(pa.x())
+                + std::to_string(pa.y())
+                + std::to_string(pb.x())
+                + std::to_string(pb.y());
+        }
     }
 
     {
