diff --git a/src/gui/text/qtextengine.cpp b/src/gui/text/qtextengine.cpp
index febc364f..78cd01e9 100644
--- a/src/gui/text/qtextengine.cpp
+++ b/src/gui/text/qtextengine.cpp
@@ -2660,7 +2660,11 @@ QTextEngine::LayoutData::LayoutData(const QString &str, void **stack_memory, qsi
 
     qsizetype space_charAttributes = sizeof(QCharAttributes) * string.size() / sizeof(void*) + 1;
     qsizetype space_logClusters = sizeof(unsigned short) * string.size() / sizeof(void*) + 1;
-    available_glyphs = (allocated - space_charAttributes - space_logClusters) * sizeof(void*) / QGlyphLayout::SpaceNeeded;
+    if (allocated > space_charAttributes + space_logClusters) {
+        available_glyphs = (allocated - space_charAttributes - space_logClusters) * sizeof(void*) / QGlyphLayout::SpaceNeeded;
+    } else {
+        available_glyphs = 0;
+    }
 
     if (available_glyphs < str.size()) {
         // need to allocate on the heap
