diff --git a/src/openglwidgets/qopenglwidget.cpp b/src/openglwidgets/qopenglwidget.cpp
index 2097da2b..aa1213a0 100644
--- a/src/openglwidgets/qopenglwidget.cpp
+++ b/src/openglwidgets/qopenglwidget.cpp
@@ -983,10 +983,6 @@ void QOpenGLWidgetPrivate::render()
 
     QOpenGLContextPrivate::get(ctx)->defaultFboRedirect = fbos[currentTargetBuffer]->handle();
 
-    f->glUseProgram(0);
-    f->glBindBuffer(GL_ARRAY_BUFFER, 0);
-    f->glEnable(GL_BLEND);
-
     q->paintGL();
     if (updateBehavior == QOpenGLWidget::NoPartialUpdate)
         invalidateFboAfterPainting();
