diff --git a/src/widgets/kernel/qapplication.cpp b/src/widgets/kernel/qapplication.cpp
index f5236d910e..5023812387 100644
--- a/src/widgets/kernel/qapplication.cpp
+++ b/src/widgets/kernel/qapplication.cpp
@@ -1875,7 +1875,7 @@ void QApplicationPrivate::setActiveWindow(QWidget* act)
             QApplicationPrivate::setFocusWidget(nullptr, Qt::ActiveWindowFocusReason);
         } else if (QApplicationPrivate::active_window) {
             QWidget *w = QApplicationPrivate::active_window->focusWidget();
-            if (w && w->isVisible() /*&& w->focusPolicy() != QWidget::NoFocus*/)
+            if (w && w->isVisibleTo(QApplicationPrivate::active_window) /*&& w->focusPolicy() != QWidget::NoFocus*/)
                 w->setFocus(Qt::ActiveWindowFocusReason);
             else {
                 w = QApplicationPrivate::focusNextPrevChild_helper(QApplicationPrivate::active_window, true);
