diff --git a/src/gui/text/qtextlayout.cpp b/src/gui/text/qtextlayout.cpp
index fd48132f5e..f9855c3bc8 100644
--- a/src/gui/text/qtextlayout.cpp
+++ b/src/gui/text/qtextlayout.cpp
@@ -2108,6 +2108,10 @@ found:
     }
 
     if (hasInlineObject && QTextDocumentPrivate::get(eng->block) != nullptr) {
+        // Patch: Attempt to fix inline objects without text in line.
+        if (!line.height()) {
+            line.setDefaultHeight(eng);
+        }
         // position top/bottom aligned inline objects
         if (maxInlineObjectHeight > line.ascent + line.descent) {
             // extend line height if required
