diff --git a/src/plugins/platforms/cocoa/qcocoafiledialoghelper.mm b/src/plugins/platforms/cocoa/qcocoafiledialoghelper.mm
index 91d76fa254..74b043e049 100644
--- a/src/plugins/platforms/cocoa/qcocoafiledialoghelper.mm
+++ b/src/plugins/platforms/cocoa/qcocoafiledialoghelper.mm
@@ -394,6 +394,10 @@ typedef QSharedPointer<QFileDialogOptions> SharedPointerFileDialogOptions;
     if (!m_helper)
         return;
 
+    // Patch: Fix crash when searching in Finder.
+    if (!path || [path isEqual:[NSNull null]])
+        return;
+
     if (!(path && path.length) || [path isEqualToString:m_currentDirectory])
         return;
 
