diff --git a/src/gui/kernel/qhighdpiscaling.cpp b/src/gui/kernel/qhighdpiscaling.cpp
index 996eed40cb..1b9a7929bf 100644
--- a/src/gui/kernel/qhighdpiscaling.cpp
+++ b/src/gui/kernel/qhighdpiscaling.cpp
@@ -632,13 +632,13 @@ qreal QHighDpiScaling::screenSubfactor(const QPlatformScreen *screen)
         if (auto qScreen = screen->screen()) {
             auto screenFactor = qScreen->property(scaleFactorProperty).toReal(&screenPropertyUsed);
             if (screenPropertyUsed)
-                factor = screenFactor;
+                factor = roundScaleFactor(screenFactor);
         }
 
         if (!screenPropertyUsed) {
             auto byNameIt = QHighDpiScaling::m_namedScreenScaleFactors.constFind(screen->name());
             if ((screenPropertyUsed = byNameIt != QHighDpiScaling::m_namedScreenScaleFactors.cend()))
-                factor = *byNameIt;
+                factor = roundScaleFactor(*byNameIt);
         }
     }
 
