diff --git a/src/plugins/platforms/cocoa/qnswindow.mm b/src/plugins/platforms/cocoa/qnswindow.mm
index 6244d5d129..7351e789be 100644
--- a/src/plugins/platforms/cocoa/qnswindow.mm
+++ b/src/plugins/platforms/cocoa/qnswindow.mm
@@ -278,14 +278,8 @@ static bool isMouseEvent(NSEvent *ev)
     // or a QWindow::backgroundColor() property. In the meantime
     // we assume that if you have translucent content, without a
     // frame then you intend to do all background drawing yourself.
-    const QWindow *window = m_platformWindow ? m_platformWindow->window() : nullptr;
-    if (!self.opaque && window) {
-        // Qt::Popup also requires clearColor - in qmacstyle
-        // we fill background using a special path with rounded corners.
-        if (window->flags().testFlag(Qt::FramelessWindowHint)
-            || (window->flags() & Qt::WindowType_Mask) == Qt::Popup)
-            return [NSColor clearColor];
-    }
+    if (!self.opaque)
+        return [NSColor clearColor];
 
     // This still allows you to have translucent content with a frame,
     // where the system background (or color set via NSWindow) will
