diff --git a/src/plugins/hardwareintegration/client/wayland-egl/main.cpp b/src/plugins/hardwareintegration/client/wayland-egl/main.cpp
index 08906355..8031132b 100644
--- a/src/plugins/hardwareintegration/client/wayland-egl/main.cpp
+++ b/src/plugins/hardwareintegration/client/wayland-egl/main.cpp
@@ -40,6 +40,10 @@
 #include <QtWaylandClient/private/qwaylandclientbufferintegrationplugin_p.h>
 #include "qwaylandeglclientbufferintegration.h"
 
+#ifdef QT_STATICPLUGIN
+extern "C" bool epoxy_load_egl(bool exit_if_fails, bool load);
+#endif
+
 QT_BEGIN_NAMESPACE
 
 namespace QtWaylandClient {
@@ -54,6 +58,12 @@ public:
 
 QWaylandClientBufferIntegration *QWaylandEglClientBufferPlugin::create(const QString&, const QStringList&)
 {
+#ifdef QT_STATICPLUGIN
+    if (!epoxy_load_egl(false, true)) {
+        return nullptr;
+    }
+#endif
+
     return new QWaylandEglClientBufferIntegration();
 }
 
