diff --git a/src/corelib/text/qharfbuzz.cpp b/src/corelib/text/qharfbuzz.cpp
index f54ce26206..caec44ee83 100644
--- a/src/corelib/text/qharfbuzz.cpp
+++ b/src/corelib/text/qharfbuzz.cpp
@@ -72,7 +72,10 @@ HB_UChar16 HB_GetMirroredChar(HB_UChar16 ch)
 
 void (*HB_Library_Resolve(const char *library, int version, const char *symbol))()
 {
-#if !QT_CONFIG(library) || defined(Q_OS_WASM)
+    // Disable dynamic library loading on Windows for harfbuzz.
+    // It is used for loading 'libthai' on Linux,
+    // but this leads to possible dll hijacking on Windows.
+#if !QT_CONFIG(library) || defined(Q_OS_WASM) || defined(Q_OS_WIN)
     Q_UNUSED(library);
     Q_UNUSED(version);
     Q_UNUSED(symbol);
