diff --git a/src/plugins/platforms/windows/qwindowseglcontext.cpp b/src/plugins/platforms/windows/qwindowseglcontext.cpp
index 4ae087dfaa..00e8242a38 100644
--- a/src/plugins/platforms/windows/qwindowseglcontext.cpp
+++ b/src/plugins/platforms/windows/qwindowseglcontext.cpp
@@ -48,6 +48,14 @@
 #  include <EGL/eglext.h>
 #endif
 
+#ifndef EGL_PLATFORM_ANGLE_DEVICE_TYPE_WARP_ANGLE
+#define EGL_PLATFORM_ANGLE_DEVICE_TYPE_WARP_ANGLE EGL_PLATFORM_ANGLE_DEVICE_TYPE_D3D_WARP_ANGLE
+#endif // EGL_PLATFORM_ANGLE_DEVICE_TYPE_WARP_ANGLE
+
+#ifndef EGL_PLATFORM_ANGLE_DEVICE_TYPE_REFERENCE_ANGLE
+#define EGL_PLATFORM_ANGLE_DEVICE_TYPE_REFERENCE_ANGLE EGL_PLATFORM_ANGLE_DEVICE_TYPE_D3D_REFERENCE_ANGLE
+#endif // EGL_PLATFORM_ANGLE_DEVICE_TYPE_REFERENCE_ANGLE
+
 QT_BEGIN_NAMESPACE
 
 /*!
