diff --git a/src/gui/painting/qplatformbackingstore.cpp b/src/gui/painting/qplatformbackingstore.cpp
index c092a7153f..5400f66ecc 100644
--- a/src/gui/painting/qplatformbackingstore.cpp
+++ b/src/gui/painting/qplatformbackingstore.cpp
@@ -102,11 +102,13 @@ public:
             QOffscreenSurface offscreenSurface;
             offscreenSurface.setFormat(context->format());
             offscreenSurface.create();
-            context->makeCurrent(&offscreenSurface);
-            if (textureId)
-                context->functions()->glDeleteTextures(1, &textureId);
-            if (blitter)
-                blitter->destroy();
+            if (context->makeCurrent(&offscreenSurface)) {
+                // Trying to fix a crash on context loss.
+                if (textureId)
+                    context->functions()->glDeleteTextures(1, &textureId);
+                if (blitter)
+                    blitter->destroy();
+            }
         }
         delete blitter;
 #endif
diff --git a/src/widgets/kernel/qopenglwidget.cpp b/src/widgets/kernel/qopenglwidget.cpp
index 24e0c46032..3cd75a00ab 100644
--- a/src/widgets/kernel/qopenglwidget.cpp
+++ b/src/widgets/kernel/qopenglwidget.cpp
@@ -647,9 +647,11 @@ void QOpenGLWidgetPaintDevice::ensureActiveTarget()
     if (!wd->initialized)
         return;
 
-    if (QOpenGLContext::currentContext() != wd->context)
+    if (QOpenGLContext::currentContext() != wd->context) {
         d->w->makeCurrent();
-    else
+        if (!wd->initialized)
+            return; // Trying to fix a crash on context loss.
+    } else
         wd->fbo->bind();
 
     if (!wd->inPaintGL)
@@ -729,7 +731,11 @@ void QOpenGLWidgetPrivate::recreateFbo()
 
     emit q->aboutToResize();
 
-    context->makeCurrent(surface);
+    if (!context->makeCurrent(surface)) {
+        // Trying to fix a crash on context loss.
+        reset();
+        return;
+    }
 
     delete fbo;
     fbo = nullptr;
@@ -770,6 +776,9 @@ void QOpenGLWidgetPrivate::beginCompose()
     if (flushPending) {
         flushPending = false;
         q->makeCurrent();
+        if (!initialized) {
+            return;
+        }
         static_cast<QOpenGLExtensions *>(context->functions())->flushShared();
     }
     hasBeenComposed = true;
@@ -812,6 +821,7 @@ void QOpenGLWidgetPrivate::initialize()
     }
     if (Q_UNLIKELY(!ctx->create())) {
         qWarning("QOpenGLWidget: Failed to create context");
+        reset();
         return;
     }
 
@@ -842,6 +852,7 @@ void QOpenGLWidgetPrivate::initialize()
 
     if (Q_UNLIKELY(!ctx->makeCurrent(surface))) {
         qWarning("QOpenGLWidget: Failed to make context current");
+        reset();
         return;
     }
 
@@ -860,6 +871,9 @@ void QOpenGLWidgetPrivate::resolveSamples()
     Q_Q(QOpenGLWidget);
     if (resolvedFbo) {
         q->makeCurrent();
+        if (!initialized) {
+            return;
+        }
         QRect rect(QPoint(0, 0), fbo->size());
         QOpenGLFramebufferObject::blitFramebuffer(resolvedFbo, rect, fbo, rect);
         flushPending = true;
@@ -893,6 +907,8 @@ void QOpenGLWidgetPrivate::render()
         return;
 
     q->makeCurrent();
+    if (!initialized)
+        return; // Trying to fix a crash on context loss.
 
     if (updateBehavior == QOpenGLWidget::NoPartialUpdate && hasBeenComposed) {
         invalidateFbo();
@@ -941,14 +957,21 @@ QImage QOpenGLWidgetPrivate::grabFramebuffer()
     if (!fbo) // could be completely offscreen, without ever getting a resize event
         recreateFbo();
 
+    if (!fbo)
+        return QImage(); // Trying to fix a crash on context loss.
+
     if (!inPaintGL)
         render();
 
     if (resolvedFbo) {
         resolveSamples();
+        if (!initialized)
+            return QImage(); // Trying to fix a crash on context loss.
         resolvedFbo->bind();
     } else {
         q->makeCurrent();
+        if (!initialized)
+            return QImage(); // Trying to fix a crash on context loss.
     }
 
     const bool hasAlpha = q->format().hasAlpha();
@@ -1158,7 +1181,12 @@ void QOpenGLWidget::makeCurrent()
     if (!d->initialized)
         return;
 
-    d->context->makeCurrent(d->surface);
+    if (!d->context->makeCurrent(d->surface)) {
+        // Trying to fix a crash on context loss.
+        // If makeCurrent() failed, that means we're not initialized any more.
+        d->initialized = false; // This prevents infinite recursion to makeCurrent().
+        d->reset();
+    }
 
     if (d->fbo) // there may not be one if we are in reset()
         d->fbo->bind();
