diff --git a/src/corelib/kernel/kernel.pri b/src/corelib/kernel/kernel.pri
index 749672c899..f08e4ebefb 100644
--- a/src/corelib/kernel/kernel.pri
+++ b/src/corelib/kernel/kernel.pri
@@ -163,6 +163,7 @@ unix|integrity {
         HEADERS += \
             kernel/qeventdispatcher_glib_p.h
         QMAKE_USE_PRIVATE += glib
+        qtConfig(dlopen): QMAKE_USE_PRIVATE += libdl
     }
 
     qtConfig(clock-gettime): QMAKE_USE_PRIVATE += librt
diff --git a/src/corelib/kernel/qeventdispatcher_glib.cpp b/src/corelib/kernel/qeventdispatcher_glib.cpp
index 92f3553247..4b85a0b1de 100644
--- a/src/corelib/kernel/qeventdispatcher_glib.cpp
+++ b/src/corelib/kernel/qeventdispatcher_glib.cpp
@@ -50,6 +50,10 @@
 
 #include <glib.h>
 
+#if QT_CONFIG(dlopen)
+#include <dlfcn.h>
+#endif
+
 QT_BEGIN_NAMESPACE
 
 struct GPollFDWithQSocketNotifier
@@ -598,6 +602,15 @@ void QEventDispatcherGlib::flush()
 
 bool QEventDispatcherGlib::versionSupported()
 {
+#if QT_CONFIG(dlopen)
+    const auto handle = dlopen("libglib-2.0.so.0", RTLD_LAZY);
+    if (!handle) {
+        Q_UNUSED(dlerror());
+        return false;
+    }
+    dlclose(handle);
+#endif
+
 #if !defined(GLIB_MAJOR_VERSION) || !defined(GLIB_MINOR_VERSION) || !defined(GLIB_MICRO_VERSION)
     return false;
 #else
