diff --git a/src/gui/kernel/qopenglcontext.cpp b/src/gui/kernel/qopenglcontext.cpp
index c25132dc734..f49d4c86e4e 100644
--- a/src/gui/kernel/qopenglcontext.cpp
+++ b/src/gui/kernel/qopenglcontext.cpp
@@ -677,8 +677,14 @@ bool QOpenGLContext::makeCurrent(QSurface *surface)
         return false;
     }
 
-    if (!d->platformGLContext->makeCurrent(surface->surfaceHandle()))
+    if (!d->platformGLContext->makeCurrent(surface->surfaceHandle())) {
+        // In this place the context could switch from isValid to !isValid.
+        // It may still be current (from the previous successful calls),
+        // so we need to make everything look as if it was not set current.
+        if (QOpenGLContext::currentContext() == this)
+            doneCurrent();
         return false;
+    }
 
     QOpenGLContextPrivate::setCurrentContext(this);
 #ifndef QT_NO_DEBUG
@@ -801,6 +807,12 @@ void QOpenGLContext::swapBuffers(QSurface *surface)
     if (surface->format().swapBehavior() == QSurfaceFormat::SingleBuffer)
         functions()->glFlush();
     d->platformGLContext->swapBuffers(surfaceHandle);
+
+    // The swapBuffers call could switch the context from isValid to !isValid.
+    // It may still be current (from the previous successful calls),
+    // so we need to make everything look as if it was not set current.
+    if (QOpenGLContext::currentContext() == this && !isValid())
+        doneCurrent();
 }
 
 /*!
