diff --git a/cmake/3rdparty/extra-cmake-modules/find-modules/FindXCB.cmake b/cmake/3rdparty/extra-cmake-modules/find-modules/FindXCB.cmake
index 26b9bf89633..0c546d09a8b 100644
--- a/cmake/3rdparty/extra-cmake-modules/find-modules/FindXCB.cmake
+++ b/cmake/3rdparty/extra-cmake-modules/find-modules/FindXCB.cmake
@@ -145,7 +145,7 @@ endforeach()
 set(XCB_XCB_component_deps)
 set(XCB_COMPOSITE_component_deps XCB XFIXES)
 set(XCB_DAMAGE_component_deps XCB XFIXES)
-set(XCB_IMAGE_component_deps XCB SHM)
+set(XCB_IMAGE_component_deps XCB SHM AUX)
 set(XCB_RENDERUTIL_component_deps XCB RENDER)
 set(XCB_XFIXES_component_deps XCB RENDER SHAPE)
 set(XCB_XVMC_component_deps XCB XV)
