diff --git a/meson.build b/meson.build
index 76ad51d3fb..fc2fe749ea 100644
--- a/meson.build
+++ b/meson.build
@@ -3377,7 +3377,7 @@ public_programs += executable(
         'systemd-detect-virt',
         'src/detect-virt/detect-virt.c',
         include_directories : includes,
-        link_with : [libshared],
+        link_with : [libshared_static, libsystemd_static, libbasic],
         install_rpath : rootpkglibdir,
         install : true)
 
