diff --git a/src/plugins/platforms/windows/qwindowscontext.cpp b/src/plugins/platforms/windows/qwindowscontext.cpp
index 38b9823d6b..eaea51d001 100644
--- a/src/plugins/platforms/windows/qwindowscontext.cpp
+++ b/src/plugins/platforms/windows/qwindowscontext.cpp
@@ -1014,7 +1014,6 @@ static inline bool isInputMessage(UINT m)
     case WM_NCMOUSELEAVE:
     case WM_SIZING:
     case WM_MOVING:
-    case WM_SYSCOMMAND:
     case WM_COMMAND:
     case WM_DWMNCRENDERINGCHANGED:
     case WM_PAINT:
