diff --git a/src/gui/platform/unix/qgenericunixthemes.cpp b/src/gui/platform/unix/qgenericunixthemes.cpp
index 86f2a266ea..49be369452 100644
--- a/src/gui/platform/unix/qgenericunixthemes.cpp
+++ b/src/gui/platform/unix/qgenericunixthemes.cpp
@@ -853,24 +853,12 @@ QStringList QGenericUnixTheme::themeNames()
     QStringList result;
     if (QGuiApplication::desktopSettingsAware()) {
         const QByteArray desktopEnvironment = QGuiApplicationPrivate::platformIntegration()->services()->desktopEnvironment();
-        QList<QByteArray> gtkBasedEnvironments;
-        gtkBasedEnvironments << "GNOME"
-                             << "X-CINNAMON"
-                             << "UNITY"
-                             << "MATE"
-                             << "XFCE"
-                             << "LXDE";
         const QList<QByteArray> desktopNames = desktopEnvironment.split(':');
         for (const QByteArray &desktopName : desktopNames) {
             if (desktopEnvironment == "KDE") {
 #if QT_CONFIG(settings)
                 result.push_back(QLatin1String(QKdeTheme::name));
 #endif
-            } else if (gtkBasedEnvironments.contains(desktopName)) {
-                // prefer the GTK3 theme implementation with native dialogs etc.
-                result.push_back(QStringLiteral("gtk3"));
-                // fallback to the generic Gnome theme if loading the GTK3 theme fails
-                result.push_back(QLatin1String(QGnomeTheme::name));
             } else {
                 // unknown, but lowercase the name (our standard practice) and
                 // remove any "x-" prefix
@@ -878,6 +866,12 @@ QStringList QGenericUnixTheme::themeNames()
                 result.push_back(s.startsWith(QLatin1String("x-")) ? s.mid(2) : s);
             }
         }
+        if (!result.contains(QLatin1String(QKdeTheme::name))) {
+            // prefer the GTK3 theme implementation with native dialogs etc.
+            result.push_back(QStringLiteral("gtk3"));
+            // fallback to the generic Gnome theme if loading the GTK3 theme fails
+            result.push_back(QLatin1String(QGnomeTheme::name));
+        }
     } // desktopSettingsAware
     result.append(QLatin1String(QGenericUnixTheme::name));
     return result;
