diff --git a/src/gui/painting/qcoregraphics.mm b/src/gui/painting/qcoregraphics.mm
index ef29f4a490..ba843b7059 100644
--- a/src/gui/painting/qcoregraphics.mm
+++ b/src/gui/painting/qcoregraphics.mm
@@ -243,32 +243,29 @@ QColor qt_mac_toQColor(CGColorRef color)
 QColor qt_mac_toQColor(const NSColor *color)
 {
     QColor qtColor;
-    switch (color.type) {
-    case NSColorTypeComponentBased: {
+    // Patch: NSColor.type was added in 10.13.
+    if (__builtin_available(macOS 10.13, *)) {
+        if (color.type == NSColorTypeComponentBased) {
         const NSColorSpace *colorSpace = [color colorSpace];
         if (colorSpace == NSColorSpace.genericRGBColorSpace
             && color.numberOfComponents == 4) { // rbga
             CGFloat components[4];
             [color getComponents:components];
             qtColor.setRgbF(components[0], components[1], components[2], components[3]);
-            break;
+            return qtColor;
         } else if (colorSpace == NSColorSpace.genericCMYKColorSpace
                    && color.numberOfComponents == 5) { // cmyk + alpha
             CGFloat components[5];
             [color getComponents:components];
             qtColor.setCmykF(components[0], components[1], components[2], components[3], components[4]);
-            break;
+            return qtColor;
+        }
         }
     }
-        Q_FALLTHROUGH();
-    default: {
         const NSColor *tmpColor = [color colorUsingColorSpace:NSColorSpace.genericRGBColorSpace];
         CGFloat red = 0, green = 0, blue = 0, alpha = 0;
         [tmpColor getRed:&red green:&green blue:&blue alpha:&alpha];
         qtColor.setRgbF(red, green, blue, alpha);
-        break;
-    }
-    }
 
     return qtColor;
 }
@@ -295,9 +292,10 @@ static bool qt_mac_isSystemColorOrInstance(const NSColor *color, NSString *color
     // We specifically do not want isKindOfClass: here
     if ([color.className isEqualToString:className]) // NSPatternColorSpace
         return true;
-    if (color.type == NSColorTypeCatalog &&
-        [color.catalogNameComponent isEqualToString:@"System"] &&
-        [color.colorNameComponent isEqualToString:colorNameComponent])
+    // Patch: NSColor.type was added in 10.13.
+    if ([color.catalogNameComponent isEqualToString:@"System"] &&
+        [color.colorNameComponent isEqualToString:colorNameComponent] &&
+        [color.colorSpaceName isEqualToString:NSNamedColorSpace])
         return true;
     return false;
 }
@@ -353,8 +351,9 @@ QBrush qt_mac_toQBrush(const NSColor *color, QPalette::ColorGroup colorGroup)
         return qtBrush;
     }
 
-    if (color.type == NSColorTypePattern) {
-        NSImage *patternImage = color.patternImage;
+    // Patch: NSColor.type was added in 10.13.
+    if (NSColor *patternColor = [color colorUsingColorSpaceName:NSPatternColorSpace]) {
+        NSImage *patternImage = patternColor.patternImage;
         const QSizeF sz(patternImage.size.width, patternImage.size.height);
         // FIXME: QBrush is not resolution independent (QTBUG-49774)
         qtBrush.setTexture(qt_mac_toQPixmap(patternImage, sz));
