diff --git a/src/gui/CMakeLists.txt b/src/gui/CMakeLists.txt
index 905afd4038..4715a8b31e 100644
--- a/src/gui/CMakeLists.txt
+++ b/src/gui/CMakeLists.txt
@@ -843,8 +843,8 @@ qt_internal_extend_target(Gui CONDITION QT_FEATURE_egl AND NOT QT_FEATURE_egl_x1
         QT_EGL_NO_X11
 )
 
-qt_internal_extend_target(Gui CONDITION QT_FEATURE_dlopen AND QT_FEATURE_egl
-    LIBRARIES
+qt_internal_extend_target(Gui CONDITION QT_FEATURE_dlopen
+    PUBLIC_LIBRARIES
         ${CMAKE_DL_LIBS}
 )
 
