diff --git a/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp b/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp
index 6bd3c9395c9..7811dc2e425 100644
--- a/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp
+++ b/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp
@@ -83,12 +83,8 @@ void QWindowsUiaMainProvider::notifyStateChange(QAccessibleStateChangeEvent *eve
 {
     if (QAccessibleInterface *accessible = event->accessibleInterface()) {
         if (event->changedStates().checked || event->changedStates().checkStateMixed) {
-            // Notifies states changes in checkboxes, switches, and checkable item view items.
-            if (accessible->role() == QAccessible::CheckBox
-                || accessible->role() == QAccessible::Switch
-                || accessible->role() == QAccessible::Cell
-                || accessible->role() == QAccessible::ListItem
-                || accessible->role() == QAccessible::TreeItem) {
+           // Notifies states changes in checkboxes.
+           if (accessible->state().checkable) {
                 if (auto provider = providerForAccessible(accessible)) {
                     long toggleState = ToggleState_Off;
                     if (accessible->state().checked)
