diff --git a/src/corelib/kernel/qcore_mac.mm b/src/corelib/kernel/qcore_mac.mm
index 266faca0ed..38a559877e 100644
--- a/src/corelib/kernel/qcore_mac.mm
+++ b/src/corelib/kernel/qcore_mac.mm
@@ -140,7 +140,8 @@ QMacAutoReleasePool::QMacAutoReleasePool()
 {
     Class trackerClass = [QMacAutoReleasePoolTracker class];
 
-#ifdef QT_DEBUG
+// Patch: Disable this debug code because it is very slow.
+#if 0
     void *poolFrame = nullptr;
     void *frame;
     if (backtrace_from_fp(__builtin_frame_address(0), &frame, 1))
