diff --git a/src/gui/kernel/qwindow.cpp b/src/gui/kernel/qwindow.cpp
index 3e7cfa25bb..41671d2d64 100644
--- a/src/gui/kernel/qwindow.cpp
+++ b/src/gui/kernel/qwindow.cpp
@@ -227,6 +227,7 @@ void QWindowPrivate::init(QScreen *targetScreen)
 
     requestedFormat = QSurfaceFormat::defaultFormat();
     devicePixelRatio = connectScreen->devicePixelRatio();
+    scaleFactor = QHighDpiScaling::factor(q);
 
     QObject::connect(q, &QWindow::screenChanged, q, [q, this](QScreen *){
         // We may have changed scaling; trigger resize event if needed,
@@ -1375,11 +1376,24 @@ bool QWindowPrivate::updateDevicePixelRatio()
     if (platformWindow)
         QHighDpiScaling::checkDevicePixelRatioRounding(platformWindow);
 
+    const qreal newScaleFactor = QHighDpiScaling::factor(q);
+
     // If there is no platform window use the associated screen's devicePixelRatio,
     // which typically is the primary screen and will be correct for single-display
     // systems (a very common case).
     const qreal newDevicePixelRatio = platformWindow ?
-        platformWindow->devicePixelRatio() * QHighDpiScaling::factor(q) : q->screen()->devicePixelRatio();
+        platformWindow->devicePixelRatio() * newScaleFactor : q->screen()->devicePixelRatio();
+
+    if (newScaleFactor != scaleFactor) {
+        scaleFactor = newScaleFactor;
+
+        if (!geometryUpdateTimer.hasExpired()) {
+            if (platformWindow && q->isTopLevel())
+                platformWindow->propagateSizeHints();
+            if (!explicitGeometry.isNull() && !(q->windowStates() & Qt::WindowFullScreen))
+                q->setGeometry(explicitGeometry);
+        }
+    }
 
     if (newDevicePixelRatio == devicePixelRatio)
         return false;
@@ -1791,6 +1805,7 @@ void QWindow::setGeometry(const QRect &rect)
 {
     Q_D(QWindow);
     d->positionAutomatic = false;
+    d->explicitGeometry = rect;
     const QRect oldRect = geometry();
     if (rect == oldRect)
         return;
@@ -2560,9 +2575,12 @@ bool QWindow::event(QEvent *ev)
         paintEvent(static_cast<QPaintEvent *>(ev));
         break;
 
-    case QEvent::Show:
+    case QEvent::Show: {
+        Q_D(QWindow);
+        d->geometryUpdateTimer.setRemainingTime(1000);
         showEvent(static_cast<QShowEvent *>(ev));
         break;
+    }
 
     case QEvent::Hide:
         hideEvent(static_cast<QHideEvent *>(ev));
diff --git a/src/gui/kernel/qwindow_p.h b/src/gui/kernel/qwindow_p.h
index 7b1aad4838..91ae5035d3 100644
--- a/src/gui/kernel/qwindow_p.h
+++ b/src/gui/kernel/qwindow_p.h
@@ -22,6 +22,7 @@
 
 #include <QtCore/private/qobject_p.h>
 #include <QtCore/qelapsedtimer.h>
+#include <QtCore/qdeadlinetimer.h>
 #include <QtCore/qxpfunctional.h>
 #include <QtGui/qicon.h>
 #include <QtGui/qpalette.h>
@@ -108,7 +109,10 @@ public:
     QString windowFilePath;
     QIcon windowIcon;
     QRect geometry;
+    QRect explicitGeometry;
+    QDeadlineTimer geometryUpdateTimer;
     qreal devicePixelRatio = 1.0;
+    qreal scaleFactor = 1.0;
     Qt::WindowStates windowState = Qt::WindowNoState;
     QWindow::Visibility visibility = QWindow::Hidden;
     bool resizeEventPending = true;
