diff --git a/src/gui/image/qiconloader.cpp b/src/gui/image/qiconloader.cpp
index 65785c406c2..8d374300f19 100644
--- a/src/gui/image/qiconloader.cpp
+++ b/src/gui/image/qiconloader.cpp
@@ -477,6 +477,7 @@ QThemeIconInfo QIconLoader::findIconHelper(const QString &themeName,
     if (info.entries.empty()) {
         const QString svgIconName = iconNameFallback + ".svg"_L1;
         const QString pngIconName = iconNameFallback + ".png"_L1;
+        auto pngs = QThemeIconEntries();
 
         // Add all relevant files
         for (int i = 0; i < contentDirs.size(); ++i) {
@@ -519,7 +520,7 @@ QThemeIconInfo QIconLoader::findIconHelper(const QString &themeName,
                     iconEntry->filename = pngPath;
                     // Notice we ensure that pixmap entries always come before
                     // scalable to preserve search order afterwards
-                    info.entries.insert(info.entries.begin(), std::move(iconEntry));
+                    pngs.push_back(std::move(iconEntry));
                 } else if (m_supportsSvg) {
                     const QString svgPath = subDir + svgIconName;
                     if (QFile::exists(svgPath)) {
@@ -532,6 +533,13 @@ QThemeIconInfo QIconLoader::findIconHelper(const QString &themeName,
             }
         }
 
+        if (!pngs.empty()) {
+            info.entries.insert(
+                info.entries.begin(),
+                std::make_move_iterator(pngs.begin()),
+                std::make_move_iterator(pngs.end()));
+        }
+
         if (!info.entries.empty()) {
             info.iconName = iconNameFallback.toString();
         }
