diff --git a/src/network/access/qnetworkreplyhttpimpl.cpp b/src/network/access/qnetworkreplyhttpimpl.cpp
index 9d338f9003..e91333bdcc 100644
--- a/src/network/access/qnetworkreplyhttpimpl.cpp
+++ b/src/network/access/qnetworkreplyhttpimpl.cpp
@@ -2174,6 +2174,8 @@ void QNetworkReplyHttpImplPrivate::finished()
 {
     Q_Q(QNetworkReplyHttpImpl);
 
+    QPointer<QNetworkReplyHttpImpl> guard = q;
+
     if (state == Finished || state == Aborted || state == WaitingForSession)
         return;
 
@@ -2203,6 +2205,8 @@ void QNetworkReplyHttpImplPrivate::finished()
     }
 #endif
 
+    if (!guard) return;
+
     // if we don't know the total size of or we received everything save the cache
     if (totalSize.isNull() || totalSize == -1 || bytesDownloaded == totalSize)
         completeCacheSave();
@@ -2212,6 +2216,8 @@ void QNetworkReplyHttpImplPrivate::finished()
     if (isHttpRedirectResponse() && errorCode == QNetworkReply::NoError)
         return;
 
+    if (!guard) return;
+
     state = Finished;
     q->setFinished(true);
 
