diff --git a/src/plugins/platformthemes/gtk3/qgtk3theme.cpp b/src/plugins/platformthemes/gtk3/qgtk3theme.cpp
index 077955eb4e..5c8a3dddf7 100644
--- a/src/plugins/platformthemes/gtk3/qgtk3theme.cpp
+++ b/src/plugins/platformthemes/gtk3/qgtk3theme.cpp
@@ -153,7 +153,7 @@ bool QGtk3Theme::usePlatformNativeDialog(DialogType type) const
     case ColorDialog:
         return true;
     case FileDialog:
-        return useNativeFileDialog();
+        return true;
     case FontDialog:
         return true;
     default:
@@ -167,8 +167,6 @@ QPlatformDialogHelper *QGtk3Theme::createPlatformDialogHelper(DialogType type) c
     case ColorDialog:
         return new QGtk3ColorDialogHelper;
     case FileDialog:
-        if (!useNativeFileDialog())
-            return nullptr;
         return new QGtk3FileDialogHelper;
     case FontDialog:
         return new QGtk3FontDialogHelper;
