diff --git a/src/gui/kernel/qguiapplication.cpp b/src/gui/kernel/qguiapplication.cpp
index b4d5914207..106930b6f4 100644
--- a/src/gui/kernel/qguiapplication.cpp
+++ b/src/gui/kernel/qguiapplication.cpp
@@ -256,7 +256,7 @@ static void initThemeHints()
 static bool checkNeedPortalSupport()
 {
 #if QT_CONFIG(dbus)
-    return !QStandardPaths::locate(QStandardPaths::RuntimeLocation, QLatin1String("flatpak-info")).isEmpty() || qEnvironmentVariableIsSet("SNAP");
+    return true;
 #else
     return false;
 #endif // QT_CONFIG(dbus)
diff --git a/src/gui/platform/unix/qgenericunixthemes.cpp b/src/gui/platform/unix/qgenericunixthemes.cpp
index 86f2a266ea..87a81c2eb9 100644
--- a/src/gui/platform/unix/qgenericunixthemes.cpp
+++ b/src/gui/platform/unix/qgenericunixthemes.cpp
@@ -853,24 +853,12 @@ QStringList QGenericUnixTheme::themeNames()
     QStringList result;
     if (QGuiApplication::desktopSettingsAware()) {
         const QByteArray desktopEnvironment = QGuiApplicationPrivate::platformIntegration()->services()->desktopEnvironment();
-        QList<QByteArray> gtkBasedEnvironments;
-        gtkBasedEnvironments << "GNOME"
-                             << "X-CINNAMON"
-                             << "UNITY"
-                             << "MATE"
-                             << "XFCE"
-                             << "LXDE";
         const QList<QByteArray> desktopNames = desktopEnvironment.split(':');
         for (const QByteArray &desktopName : desktopNames) {
             if (desktopEnvironment == "KDE") {
 #if QT_CONFIG(settings)
                 result.push_back(QLatin1String(QKdeTheme::name));
 #endif
-            } else if (gtkBasedEnvironments.contains(desktopName)) {
-                // prefer the GTK3 theme implementation with native dialogs etc.
-                result.push_back(QStringLiteral("gtk3"));
-                // fallback to the generic Gnome theme if loading the GTK3 theme fails
-                result.push_back(QLatin1String(QGnomeTheme::name));
             } else {
                 // unknown, but lowercase the name (our standard practice) and
                 // remove any "x-" prefix
@@ -878,6 +866,10 @@ QStringList QGenericUnixTheme::themeNames()
                 result.push_back(s.startsWith(QLatin1String("x-")) ? s.mid(2) : s);
             }
         }
+        // prefer the GTK3 theme implementation with native dialogs etc.
+        result.push_back(QStringLiteral("gtk3"));
+        // fallback to the generic Gnome theme if loading the GTK3 theme fails
+        result.push_back(QLatin1String(QGnomeTheme::name));
     } // desktopSettingsAware
     result.append(QLatin1String(QGenericUnixTheme::name));
     return result;
