diff --git a/src/gui/kernel/qplatformscreen.cpp b/src/gui/kernel/qplatformscreen.cpp
index 0649139455..f7aaab3e2f 100644
--- a/src/gui/kernel/qplatformscreen.cpp
+++ b/src/gui/kernel/qplatformscreen.cpp
@@ -191,8 +191,12 @@ QDpi QPlatformScreen::logicalDpi() const
 // which accounts for QT_FONT_DPI.
 QPair<qreal, qreal> QPlatformScreen::overrideDpi(const QPair<qreal, qreal> &in)
 {
+#ifndef QT_NO_HIGHDPISCALING
     static const int overrideDpi = qEnvironmentVariableIntValue("QT_FONT_DPI");
     return overrideDpi > 0 ?  QDpi(overrideDpi, overrideDpi) : in;
+#else
+    return in;
+#endif
 }
 
 /*!
