diff --git a/src/gui/text/qtextlayout.cpp b/src/gui/text/qtextlayout.cpp
index 03eaa9c726..bab573b625 100644
--- a/src/gui/text/qtextlayout.cpp
+++ b/src/gui/text/qtextlayout.cpp
@@ -1309,12 +1309,15 @@ void QTextLayout::drawCursor(QPainter *p, const QPointF &pos, int cursorPosition
                 si = &d->layoutData->items[itm];
             }
         }
+        // Patch: Use line geometry, otherwise smiles get glitching cursor.
+#if 0
         if (si->ascent >= 0)
             base = si->ascent;
         if (si->descent == 0)
             descent = si->descent;
         else if (si->descent > 0 && si->descent < descent)
             cursorDescent = si->descent;
+#endif
         rightToLeft = si->analysis.bidiLevel % 2;
     }
     qreal y = position.y() + (sl.y + sl.base() + sl.descent - base - descent).toReal();
