diff --git a/src/client/qwaylandwindow.cpp b/src/client/qwaylandwindow.cpp
index f44067a0..903a70f7 100644
--- a/src/client/qwaylandwindow.cpp
+++ b/src/client/qwaylandwindow.cpp
@@ -51,7 +51,7 @@ QWaylandWindow::QWaylandWindow(QWindow *window, QWaylandDisplay *display)
     , mDisplay(display)
     , mSurfaceLock(QReadWriteLock::Recursive)
     , mShellIntegration(display->shellIntegration())
-    , mResizeAfterSwap(qEnvironmentVariableIsSet("QT_WAYLAND_RESIZE_AFTER_SWAP"))
+    , mResizeAfterSwap(true)
 {
     {
         bool ok;
