diff --git a/src/plugins/platforms/xcb/qxcbconnection.h b/src/plugins/platforms/xcb/qxcbconnection.h
index 24e684866d..356fbe46f0 100644
--- a/src/plugins/platforms/xcb/qxcbconnection.h
+++ b/src/plugins/platforms/xcb/qxcbconnection.h
@@ -158,6 +158,7 @@ public:
     xcb_window_t selectionOwner(xcb_atom_t atom) const;
     xcb_window_t qtSelectionOwner();
 
+    void resetButtonState() { m_buttonState = Qt::NoButton; }
     void setButtonState(Qt::MouseButton button, bool down);
     Qt::MouseButtons buttonState() const { return m_buttonState; }
     Qt::MouseButton button() const { return m_button; }
diff --git a/src/plugins/platforms/xcb/qxcbwindow.cpp b/src/plugins/platforms/xcb/qxcbwindow.cpp
index 584f1e5406..895744b3dd 100644
--- a/src/plugins/platforms/xcb/qxcbwindow.cpp
+++ b/src/plugins/platforms/xcb/qxcbwindow.cpp
@@ -22,6 +22,7 @@
 #include "qxcbimage.h"
 #include "qxcbnativeinterface.h"
 #include "qxcbsystemtraytracker.h"
+#include "qxcbcursor.h"
 
 #include <qpa/qplatformintegration.h>
 #include <qpa/qplatformcursor.h>
@@ -2383,7 +2384,23 @@ void QXcbWindow::doStartSystemMoveResize(const QPoint &globalPos, int edges)
                    XCB_EVENT_MASK_SUBSTRUCTURE_REDIRECT | XCB_EVENT_MASK_SUBSTRUCTURE_NOTIFY,
                    (const char *)&xev);
 
+    // the pointer is ungrabbed, reset the button state
+    connection()->setMouseGrabber(nullptr);
+    connection()->resetButtonState();
     connection()->setDuringSystemMoveResize(true);
+
+    QPoint cursorPos;
+    int stateMask = 0;
+    QXcbCursor::queryPointer(connection(), nullptr, &cursorPos, &stateMask);
+
+    QWindowSystemInterface::handleMouseEvent(
+        window(),
+        mapFromGlobal(cursorPos),
+        cursorPos,
+        connection()->buttonState(),
+        connection()->button(),
+        QEvent::MouseButtonRelease,
+        connection()->keyboard()->translateModifiers(stateMask));
 }
 
 // Sends an XEmbed message.
