diff --git a/src/plugins/platforms/cocoa/qcocoawindow.mm b/src/plugins/platforms/cocoa/qcocoawindow.mm
index 24e98f5bac..454beed5fe 100644
--- a/src/plugins/platforms/cocoa/qcocoawindow.mm
+++ b/src/plugins/platforms/cocoa/qcocoawindow.mm
@@ -1722,10 +1722,12 @@ void QCocoaWindow::setWindowCursor(NSCursor *cursor)
 void QCocoaWindow::registerTouch(bool enable)
 {
     m_registerTouchCount += enable ? 1 : -1;
-    if (enable && m_registerTouchCount == 1)
-        m_view.allowedTouchTypes |= NSTouchTypeMaskIndirect;
-    else if (m_registerTouchCount == 0)
-        m_view.allowedTouchTypes &= ~NSTouchTypeMaskIndirect;
+    if (__builtin_available(macOS 10.12.2, *)) {
+        if (enable && m_registerTouchCount == 1)
+            m_view.allowedTouchTypes |= NSTouchTypeMaskIndirect;
+        else if (m_registerTouchCount == 0)
+            m_view.allowedTouchTypes &= ~NSTouchTypeMaskIndirect;
+    }
 }
 
 void QCocoaWindow::setContentBorderThickness(int topThickness, int bottomThickness)
