diff --git a/src/plugins/platforms/wayland/qwaylandcursorsurface_p.h b/src/plugins/platforms/wayland/qwaylandcursorsurface_p.h
index 6936b246ff8..d20ae664b54 100644
--- a/src/plugins/platforms/wayland/qwaylandcursorsurface_p.h
+++ b/src/plugins/platforms/wayland/qwaylandcursorsurface_p.h
@@ -62,7 +62,7 @@ public:
     {
         int scale = 0;
         for (auto *screen : m_screens)
-            scale = qMax(scale, screen->scale());
+            scale = qMax(scale, qCeil(screen->scale()));
         return scale;
     }
 
diff --git a/src/plugins/platforms/wayland/qwaylandscreen.cpp b/src/plugins/platforms/wayland/qwaylandscreen.cpp
index a005812e5ea..1fbd9c72af5 100644
--- a/src/plugins/platforms/wayland/qwaylandscreen.cpp
+++ b/src/plugins/platforms/wayland/qwaylandscreen.cpp
@@ -184,14 +184,16 @@ Qt::ScreenOrientation QWaylandScreen::orientation() const
     return m_orientation;
 }
 
-int QWaylandScreen::scale() const
+qreal QWaylandScreen::scale() const
 {
-    return mScale;
+    return devicePixelRatio();
 }
 
 qreal QWaylandScreen::devicePixelRatio() const
 {
-    return qreal(mScale);
+    return (mWaylandDisplay->fractionalScaleManager() && std::fmod(mFractionalScale, qreal(1)) != 0)
+        ? mFractionalScale
+        : qreal(mScale);
 }
 
 qreal QWaylandScreen::refreshRate() const
@@ -352,7 +354,6 @@ void QWaylandScreen::updateOutputProperties()
             m_orientation = newOrientation;
             QWindowSystemInterface::handleScreenOrientationChange(screen(), m_orientation);
         }
-        mTransform = -1;
     }
 
     QWindowSystemInterface::handleScreenRefreshRateChange(screen(), refreshRate());
@@ -407,7 +408,26 @@ void QWaylandScreen::updateXdgOutputProperties()
 {
     Q_ASSERT(mInitialized);
     Q_ASSERT(zxdg_output_v1::isInitialized());
+    bool rotated = false;
+    switch (mTransform) {
+    case WL_OUTPUT_TRANSFORM_90:
+    case WL_OUTPUT_TRANSFORM_270:
+    case WL_OUTPUT_TRANSFORM_FLIPPED_90:
+    case WL_OUTPUT_TRANSFORM_FLIPPED_270:
+        rotated = true;
+        break;
+    }
+    mFractionalScale = mGeometry.isValid() && mXdgGeometry.isValid()
+        ? std::round((rotated
+        ? qreal(mGeometry.width()) / qreal(mXdgGeometry.height())
+            + qreal(mGeometry.height()) / qreal(mXdgGeometry.width())
+        : qreal(mGeometry.width()) / qreal(mXdgGeometry.width())
+            + qreal(mGeometry.height()) / qreal(mXdgGeometry.height()))
+        * qreal(0.5) * qreal(120)) / qreal(120)
+        : qreal(1);
+    const QDpi dpi = logicalDpi();
     QWindowSystemInterface::handleScreenGeometryChange(screen(), geometry(), availableGeometry());
+    QWindowSystemInterface::handleScreenLogicalDotsPerInchChange(screen(), dpi.first, dpi.second);
 }
 
 } // namespace QtWaylandClient
diff --git a/src/plugins/platforms/wayland/qwaylandscreen_p.h b/src/plugins/platforms/wayland/qwaylandscreen_p.h
index 15cb8c0a16b..51d337e83e0 100644
--- a/src/plugins/platforms/wayland/qwaylandscreen_p.h
+++ b/src/plugins/platforms/wayland/qwaylandscreen_p.h
@@ -65,7 +65,7 @@ public:
     QList<QPlatformScreen *> virtualSiblings() const override;
 
     Qt::ScreenOrientation orientation() const override;
-    int scale() const;
+    qreal scale() const;
     qreal devicePixelRatio() const override;
     qreal refreshRate() const override;
 
@@ -124,6 +124,7 @@ protected:
     QRect mGeometry;
     QRect mXdgGeometry;
     int mScale = 1;
+    qreal mFractionalScale = 1;
     int mDepth = 32;
     int mRefreshRate = 60000;
     int mSubpixel = -1;
diff --git a/src/plugins/platforms/wayland/qwaylandwindow.cpp b/src/plugins/platforms/wayland/qwaylandwindow.cpp
index 7a83eb2d58e..b3f854d07ff 100644
--- a/src/plugins/platforms/wayland/qwaylandwindow.cpp
+++ b/src/plugins/platforms/wayland/qwaylandwindow.cpp
@@ -31,6 +31,7 @@
 #include <QGuiApplication>
 #include <qpa/qwindowsysteminterface.h>
 #include <QtGui/private/qguiapplication_p.h>
+#include <QtGui/private/qhighdpiscaling_p.h>
 #include <QtGui/private/qwindow_p.h>
 
 #include <QtCore/QDebug>
@@ -58,6 +59,7 @@ QWaylandWindow::QWaylandWindow(QWindow *window, QWaylandDisplay *display)
     , mDisplay(display)
     , mSurfaceLock(QReadWriteLock::Recursive)
     , mShellIntegration(display->shellIntegration())
+    , mGeometry(window->geometry())
     , mSurfaceFormat(window->requestedFormat())
 {
     {
@@ -236,7 +238,7 @@ void QWaylandWindow::initializeWlSurface(bool colorSpace)
     }
     emit wlSurfaceCreated();
 
-    if (mDisplay->fractionalScaleManager() && qApp->highDpiScaleFactorRoundingPolicy() == Qt::HighDpiScaleFactorRoundingPolicy::PassThrough) {
+    if (mDisplay->fractionalScaleManager()) {
         mFractionalScale.reset(new QWaylandFractionalScale(mDisplay->fractionalScaleManager()->get_fractional_scale(mSurface->object())));
 
         connect(mFractionalScale.data(), &QWaylandFractionalScale::preferredScaleChanged,
@@ -460,14 +462,29 @@ void QWaylandWindow::setWindowIcon(const QIcon &icon)
         mShellSurface->setIcon(icon);
 }
 
+bool QWaylandWindow::sizedByCompositor() const
+{
+    return (windowStates() & (Qt::WindowMaximized | Qt::WindowFullScreen))
+        || toplevelWindowTilingStates() != WindowNoState;
+}
+
 QRect QWaylandWindow::defaultGeometry() const
 {
     return QRect(QPoint(), QSize(500,500));
 }
 
+QRect QWaylandWindow::geometry() const
+{
+    if (sizedByCompositor() || qFuzzyCompare(mGeometryScale, QHighDpiScaling::factor(window())))
+        return QPlatformWindow::geometry();
+    return QHighDpi::toNativeWindowGeometry(mGeometry, window());
+}
+
 void QWaylandWindow::setGeometry_helper(const QRect &rect)
 {
     QPlatformWindow::setGeometry(rect);
+    mGeometry = QHighDpi::fromNativeWindowGeometry(rect, window());
+    mGeometryScale = QHighDpiScaling::factor(window());
     if (mViewport)
         updateViewport();
 
@@ -915,10 +932,11 @@ QMargins QWaylandWindow::clientSideMargins() const
 }
 
 void QWaylandWindow::setCustomMargins(const QMargins &margins) {
+    const auto factor = QHighDpiScaling::factor(window());
     const QMargins oldMargins = mCustomMargins;
     mCustomMargins = margins;
     propagateSizeHints();
-    setGeometry(geometry().marginsRemoved(oldMargins).marginsAdded(margins));
+    setGeometry(geometry().marginsRemoved(oldMargins * factor).marginsAdded(margins * factor));
 }
 
 /*!
@@ -937,7 +955,7 @@ QMargins QWaylandWindow::windowContentMargins() const
         shadowMargins = mWindowDecoration->margins(QWaylandAbstractDecoration::ShadowsOnly);
 
     if (!mCustomMargins.isNull())
-        shadowMargins += mCustomMargins;
+        shadowMargins += mCustomMargins * QHighDpiScaling::factor(window());
 
     return shadowMargins;
 }
@@ -1502,6 +1520,10 @@ void QWaylandWindow::handleScreensChanged()
 {
     QPlatformScreen *newScreen = calculateScreenFromSurfaceEvents();
 
+    const auto scaleGuard = qScopeGuard([&] {
+        updateScale();
+    });
+
     if (!newScreen || newScreen->screen() == window()->screen())
         return;
 
@@ -1515,12 +1537,14 @@ void QWaylandWindow::handleScreensChanged()
         setGeometry(geometry);
     }
 
-    updateScale();
     updateBufferTransform();
 }
 
 void QWaylandWindow::updateScale()
 {
+    if (QHighDpiScaling::isActive() && mSurface && mSurface->m_screens.size() > 1)
+        return;
+
     if (mFractionalScale) {
         qreal preferredScale = mFractionalScale->preferredScale().value_or(1.0);
         preferredScale = std::max<qreal>(1.0, preferredScale);
@@ -1536,7 +1560,7 @@ void QWaylandWindow::updateScale()
         return;
     }
 
-    int scale = screen()->isPlaceholder() ? 1 : static_cast<QWaylandScreen *>(screen())->scale();
+    qreal scale = screen()->isPlaceholder() ? qreal(1) : static_cast<QWaylandScreen *>(screen())->scale();
     setScale(scale);
 }
 
@@ -1544,7 +1568,9 @@ void QWaylandWindow::setScale(qreal newScale)
 {
     if (mScale.has_value() && qFuzzyCompare(mScale.value(), newScale))
         return;
+    const auto oldFactor = QHighDpiScaling::factor(window());
     mScale = newScale;
+    const auto factor = QHighDpiScaling::factor(window());
 
     if (mSurface) {
         if (mViewport)
@@ -1555,11 +1581,20 @@ void QWaylandWindow::setScale(qreal newScale)
     ensureSize();
 
     QWindowSystemInterface::handleWindowDevicePixelRatioChanged<QWindowSystemInterface::SynchronousDelivery>(window());
+    propagateSizeHints();
+    setGeometry(sizedByCompositor()
+        ? geometry().marginsRemoved(mCustomMargins * oldFactor).marginsAdded(mCustomMargins * factor)
+        : geometry());
     if (isExposed()) {
         // redraw at the new DPR
         window()->requestUpdate();
         sendExposeEvent(QRect(QPoint(), geometry().size()));
     }
+
+    for (const auto &popup : std::as_const(mChildPopups)) {
+        if (popup && popup->mShellSurface && popup->mExposed)
+            popup->mShellSurface->setWindowPosition(QHighDpi::toNativeGlobalPosition(popup->window()->position(), popup->window()));
+    }
 }
 
 #if QT_CONFIG(cursor)
@@ -1653,7 +1688,7 @@ qreal QWaylandWindow::scale() const
 
 qreal QWaylandWindow::devicePixelRatio() const
 {
-    return mScale.value_or(waylandScreen() ? waylandScreen()->scale() : 1);
+    return mScale.value_or(waylandScreen() ? waylandScreen()->scale() : qreal(1));
 }
 
 bool QWaylandWindow::setMouseGrabEnabled(bool grab)
@@ -1970,13 +2005,13 @@ QNativeInterface::Private::QWaylandWindow::WindowType QWaylandWindow::extendedWi
 void QWaylandWindow::setParentControlGeometry(const QRect &parentControlGeometry) {
     m_popupInfo.parentControlGeometry = parentControlGeometry;
     if (mExposed) {
-        mShellSurface->setWindowPosition(window()->position());
+        mShellSurface->setWindowPosition(QHighDpi::toNativeGlobalPosition(window()->position(), window()));
     }
 }
 
 QRect QWaylandWindow::parentControlGeometry() const
 {
-    return m_popupInfo.parentControlGeometry;
+    return QHighDpi::scale(m_popupInfo.parentControlGeometry, QHighDpiScaling::factor(transientParent()->window()));
 }
 
 }
diff --git a/src/plugins/platforms/wayland/qwaylandwindow_p.h b/src/plugins/platforms/wayland/qwaylandwindow_p.h
index 670a7cba6d4..898cd0d79e4 100644
--- a/src/plugins/platforms/wayland/qwaylandwindow_p.h
+++ b/src/plugins/platforms/wayland/qwaylandwindow_p.h
@@ -102,6 +102,7 @@ public:
     inline QIcon windowIcon() const;
     void setWindowIcon(const QIcon &icon) override;
 
+    QRect geometry() const override;
     void setGeometry(const QRect &rect) override;
 
     bool allowsIndependentThreadedRendering() const override;
@@ -336,6 +337,8 @@ protected:
     QVariantMap m_properties;
 
     QPoint mOffset;
+    QRect mGeometry;
+    qreal mGeometryScale = 0;
     std::optional<qreal> mScale = std::nullopt;
 
     QString mWindowTitle;
@@ -384,6 +387,7 @@ private:
     void updateInputRegion();
     void updateViewport();
     bool calculateExposure() const;
+    bool sizedByCompositor() const;
     void setPendingImageDescription();
 
     void handleMouseEventWithDecoration(QWaylandInputDevice *inputDevice, const QWaylandPointerEvent &e);
