diff --git a/src/client/qwaylandinputdevice.cpp b/src/client/qwaylandinputdevice.cpp
index a4f8757e..68938a4a 100644
--- a/src/client/qwaylandinputdevice.cpp
+++ b/src/client/qwaylandinputdevice.cpp
@@ -216,7 +216,7 @@ public:
     {
         int scale = 0;
         for (auto *screen : m_screens)
-            scale = qMax(scale, screen->scale());
+            scale = qMax(scale, qCeil(screen->scale()));
         return scale;
     }
 
diff --git a/src/client/qwaylandscreen.cpp b/src/client/qwaylandscreen.cpp
index 7c4f1bdf..f3433f4a 100644
--- a/src/client/qwaylandscreen.cpp
+++ b/src/client/qwaylandscreen.cpp
@@ -168,14 +168,16 @@ Qt::ScreenOrientation QWaylandScreen::orientation() const
     return m_orientation;
 }
 
-int QWaylandScreen::scale() const
+qreal QWaylandScreen::scale() const
 {
-    return mScale;
+    return devicePixelRatio();
 }
 
 qreal QWaylandScreen::devicePixelRatio() const
 {
-    return qreal(mScale);
+    return (mWaylandDisplay->fractionalScaleManager() && std::fmod(mFractionalScale, qreal(1)) != 0)
+        ? mFractionalScale
+        : qreal(mScale);
 }
 
 qreal QWaylandScreen::refreshRate() const
@@ -324,7 +326,12 @@ void QWaylandScreen::zxdg_output_v1_name(const QString &name)
 void QWaylandScreen::updateXdgOutputProperties()
 {
     Q_ASSERT(zxdg_output_v1::isInitialized());
+    mFractionalScale = std::round((qreal(mGeometry.width()) / qreal(mXdgGeometry.width())
+        + qreal(mGeometry.height()) / qreal(mXdgGeometry.height()))
+        * qreal(0.5) * qreal(100)) / qreal(100);
+    const QDpi dpi = logicalDpi();
     QWindowSystemInterface::handleScreenGeometryChange(screen(), geometry(), geometry());
+    QWindowSystemInterface::handleScreenLogicalDotsPerInchChange(screen(), dpi.first, dpi.second);
 }
 
 } // namespace QtWaylandClient
diff --git a/src/client/qwaylandscreen_p.h b/src/client/qwaylandscreen_p.h
index 4499f0ee..f1e761e7 100644
--- a/src/client/qwaylandscreen_p.h
+++ b/src/client/qwaylandscreen_p.h
@@ -63,7 +63,7 @@ public:
     QList<QPlatformScreen *> virtualSiblings() const override;
 
     Qt::ScreenOrientation orientation() const override;
-    int scale() const;
+    qreal scale() const;
     qreal devicePixelRatio() const override;
     qreal refreshRate() const override;
 
@@ -115,6 +115,7 @@ protected:
     QRect mGeometry;
     QRect mXdgGeometry;
     int mScale = 1;
+    qreal mFractionalScale = 1;
     int mDepth = 32;
     int mRefreshRate = 60000;
     int mTransform = -1;
diff --git a/src/client/qwaylandwindow.cpp b/src/client/qwaylandwindow.cpp
index b81e00ef..de11b82a 100644
--- a/src/client/qwaylandwindow.cpp
+++ b/src/client/qwaylandwindow.cpp
@@ -56,7 +56,7 @@ QWaylandWindow::QWaylandWindow(QWindow *window, QWaylandDisplay *display)
             mFrameCallbackTimeout = frameCallbackTimeout;
     }
 
-    mScale = waylandScreen() ? waylandScreen()->scale() : 1; // fallback to 1 if we don't have a real screen
+    mScale = waylandScreen() ? waylandScreen()->scale() : qreal(1); // fallback to 1 if we don't have a real screen
 
     static WId id = 1;
     mWindowId = id++;
@@ -102,7 +102,7 @@ void QWaylandWindow::initWindow()
         initializeWlSurface();
     }
 
-    if (mDisplay->fractionalScaleManager() && qApp->highDpiScaleFactorRoundingPolicy() == Qt::HighDpiScaleFactorRoundingPolicy::PassThrough) {
+    if (mDisplay->fractionalScaleManager()) {
         mFractionalScale.reset(new QWaylandFractionalScale(mDisplay->fractionalScaleManager()->get_fractional_scale(mSurface->object())));
 
         connect(mFractionalScale.data(), &QWaylandFractionalScale::preferredScaleChanged, this, [this](qreal preferredScale) {
@@ -1395,7 +1395,7 @@ void QWaylandWindow::handleScreensChanged()
     if (mFractionalScale)
         return;
 
-    int scale = mLastReportedScreen->isPlaceholder() ? 1 : static_cast<QWaylandScreen *>(mLastReportedScreen)->scale();
+    qreal scale = mLastReportedScreen->isPlaceholder() ? qreal(1) : static_cast<QWaylandScreen *>(mLastReportedScreen)->scale();
 
     if (scale != mScale) {
         mScale = scale;
@@ -1413,7 +1413,7 @@ void QWaylandWindow::handleScreensChanged()
 #if QT_CONFIG(cursor)
 void QWaylandWindow::setMouseCursor(QWaylandInputDevice *device, const QCursor &cursor)
 {
-    int fallbackBufferScale = int(devicePixelRatio());
+    int fallbackBufferScale = qCeil(devicePixelRatio());
     device->setCursor(&cursor, {}, fallbackBufferScale);
 }
 
