diff --git a/src/client/qwaylandtextinputv1.cpp b/src/client/qwaylandtextinputv1.cpp
index da8bbae7..3b45322b 100644
--- a/src/client/qwaylandtextinputv1.cpp
+++ b/src/client/qwaylandtextinputv1.cpp
@@ -10,6 +10,7 @@
 #include <QtCore/qloggingcategory.h>
 #include <QtGui/QGuiApplication>
 #include <QtGui/private/qguiapplication_p.h>
+#include <QtGui/private/qhighdpiscaling_p.h>
 #include <QtGui/qpa/qplatformintegration.h>
 #include <QtGui/qevent.h>
 #include <QtGui/qwindow.h>
@@ -127,8 +128,9 @@ void QWaylandTextInputv1::updateState(Qt::InputMethodQueries queries, uint32_t f
     if (queries & Qt::ImCursorRectangle) {
         const QRect &cRect = event.value(Qt::ImCursorRectangle).toRect();
         const QRect &windowRect = QGuiApplication::inputMethod()->inputItemTransform().mapRect(cRect);
-        const QMargins margins = window->frameMargins();
-        const QRect &surfaceRect = windowRect.translated(margins.left(), margins.top());
+        const QRect &nativeRect = QHighDpi::toNativePixels(windowRect, QGuiApplication::focusWindow());
+        const QMargins margins = window->clientSideMargins();
+        const QRect &surfaceRect = nativeRect.translated(margins.left(), margins.top());
         set_cursor_rectangle(surfaceRect.x(), surfaceRect.y(), surfaceRect.width(), surfaceRect.height());
     }
 
diff --git a/src/client/qwaylandtextinputv2.cpp b/src/client/qwaylandtextinputv2.cpp
index a799793e..ef6ad4be 100644
--- a/src/client/qwaylandtextinputv2.cpp
+++ b/src/client/qwaylandtextinputv2.cpp
@@ -11,6 +11,7 @@
 #include <QtCore/qloggingcategory.h>
 #include <QtGui/QGuiApplication>
 #include <QtGui/private/qguiapplication_p.h>
+#include <QtGui/private/qhighdpiscaling_p.h>
 #include <QtGui/qpa/qplatformintegration.h>
 #include <QtGui/qevent.h>
 #include <QtGui/qwindow.h>
@@ -133,8 +134,9 @@ void QWaylandTextInputv2::updateState(Qt::InputMethodQueries queries, uint32_t f
     if (queries & Qt::ImCursorRectangle) {
         const QRect &cRect = event.value(Qt::ImCursorRectangle).toRect();
         const QRect &windowRect = QGuiApplication::inputMethod()->inputItemTransform().mapRect(cRect);
-        const QMargins margins = window->frameMargins();
-        const QRect &surfaceRect = windowRect.translated(margins.left(), margins.top());
+        const QRect &nativeRect = QHighDpi::toNativePixels(windowRect, QGuiApplication::focusWindow());
+        const QMargins margins = window->clientSideMargins();
+        const QRect &surfaceRect = nativeRect.translated(margins.left(), margins.top());
         set_cursor_rectangle(surfaceRect.x(), surfaceRect.y(), surfaceRect.width(), surfaceRect.height());
     }
 
diff --git a/src/client/qwaylandtextinputv4.cpp b/src/client/qwaylandtextinputv4.cpp
index 6cf01613..79c9cacf 100644
--- a/src/client/qwaylandtextinputv4.cpp
+++ b/src/client/qwaylandtextinputv4.cpp
@@ -8,6 +8,7 @@
 
 #include <QtCore/qloggingcategory.h>
 #include <QtGui/qguiapplication.h>
+#include <QtGui/private/qhighdpiscaling_p.h>
 #include <QtGui/qevent.h>
 #include <QtGui/qwindow.h>
 #include <QTextCharFormat>
@@ -239,8 +240,9 @@ void QWaylandTextInputv4::updateState(Qt::InputMethodQueries queries, uint32_t f
     if (queries & Qt::ImCursorRectangle) {
         const QRect &cRect = event.value(Qt::ImCursorRectangle).toRect();
         const QRect &windowRect = QGuiApplication::inputMethod()->inputItemTransform().mapRect(cRect);
-        const QMargins margins = window->frameMargins();
-        const QRect &surfaceRect = windowRect.translated(margins.left(), margins.top());
+        const QRect &nativeRect = QHighDpi::toNativePixels(windowRect, QGuiApplication::focusWindow());
+        const QMargins margins = window->clientSideMargins();
+        const QRect &surfaceRect = nativeRect.translated(margins.left(), margins.top());
         if (surfaceRect != m_cursorRect) {
             set_cursor_rectangle(surfaceRect.x(), surfaceRect.y(), surfaceRect.width(), surfaceRect.height());
             m_cursorRect = surfaceRect;
