diff --git a/BUILD.gn b/BUILD.gn
index c4b555594..e9bc90ace 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -642,6 +642,10 @@ angle_source_set("angle_version") {
 config("angle_backend_config") {
   defines = []
 
+  if (angle_default_d3d9) {
+    defines += [ "ANGLE_DEFAULT_D3D11=0" ]
+  }
+
   if (angle_delegate_workers) {
     defines += [ "ANGLE_DELEGATE_WORKERS" ]
   }
diff --git a/gni/angle.gni b/gni/angle.gni
index f0dbf9db8..3bafe80b9 100644
--- a/gni/angle.gni
+++ b/gni/angle.gni
@@ -184,6 +184,10 @@ declare_args() {
   angle_has_histograms = angle_has_build
 }
 
+declare_args() {
+  angle_default_d3d9 = false
+}
+
 if (!defined(angle_zlib_compression_utils_dir)) {
   angle_zlib_compression_utils_dir = "//third_party/zlib/google"
 }
