diff --git a/src/gui/kernel/qguiapplication.cpp b/src/gui/kernel/qguiapplication.cpp
index c7ff2a6dacd..f88d6fe8deb 100644
--- a/src/gui/kernel/qguiapplication.cpp
+++ b/src/gui/kernel/qguiapplication.cpp
@@ -2548,11 +2548,6 @@ void QGuiApplicationPrivate::processWindowScreenChangedEvent(QWindowSystemInterf
             else // Fall back to default behavior, and try to find some appropriate screen
                 topLevelWindow->setScreen(nullptr);
         }
-        // we may have changed scaling, so trigger resize event if needed
-        if (window->handle()) {
-            QWindowSystemInterfacePrivate::GeometryChangeEvent gce(window, QHighDpi::fromNativePixels(window->handle()->geometry(), window));
-            processGeometryChangeEvent(&gce);
-        }
     }
 }
 
@@ -3209,6 +3204,10 @@ void QGuiApplicationPrivate::processScreenLogicalDotsPerInchChange(QWindowSystem
     emit s->logicalDotsPerInchChanged(s->logicalDotsPerInch());
     s->d_func()->updateGeometriesWithSignals();
 
+    for (QWindow *window : QGuiApplication::allWindows())
+        if (window->screen() == e->screen)
+            window->screenChanged(e->screen);
+
     resetCachedDevicePixelRatio();
 }
 
diff --git a/src/gui/kernel/qhighdpiscaling_p.h b/src/gui/kernel/qhighdpiscaling_p.h
index 5b02c5d646a..630ee0350db 100644
--- a/src/gui/kernel/qhighdpiscaling_p.h
+++ b/src/gui/kernel/qhighdpiscaling_p.h
@@ -187,7 +187,7 @@ inline QMargins scale(const QMargins &margins, qreal scaleFactor, QPoint origin
 template <typename T>
 QVector<T> scale(const QVector<T> &vector, qreal scaleFactor, QPoint origin = QPoint(0, 0))
 {
-    if (!QHighDpiScaling::isActive())
+    if (qFuzzyCompare(scaleFactor, qreal(1)))
         return vector;
 
     QVector<T> scaled;
@@ -199,7 +199,7 @@ QVector<T> scale(const QVector<T> &vector, qreal scaleFactor, QPoint origin = QP
 
 inline QRegion scale(const QRegion &region, qreal scaleFactor, QPoint origin = QPoint(0, 0))
 {
-    if (!QHighDpiScaling::isActive())
+    if (qFuzzyCompare(scaleFactor, qreal(1)))
         return region;
 
     QRegion scaled;
diff --git a/src/gui/kernel/qwindow.cpp b/src/gui/kernel/qwindow.cpp
index fa3a0243a64..a70555cae7c 100644
--- a/src/gui/kernel/qwindow.cpp
+++ b/src/gui/kernel/qwindow.cpp
@@ -244,6 +244,13 @@ void QWindowPrivate::init(QScreen *targetScreen)
     QGuiApplicationPrivate::window_list.prepend(q);
 
     requestedFormat = QSurfaceFormat::defaultFormat();
+
+    QObject::connect(q, &QWindow::screenChanged, q, [q, this](QScreen *){
+        if (q->handle()) {
+            QWindowSystemInterfacePrivate::GeometryChangeEvent gce(q, QHighDpi::fromNativePixels(q->handle()->geometry(), q));
+            QGuiApplicationPrivate::processGeometryChangeEvent(&gce);
+        }
+    });
 }
 
 /*!
@@ -523,6 +530,13 @@ void QWindowPrivate::create(bool recursive, WId nativeHandle)
     if (q->parent())
         q->parent()->create();
 
+    // QPlatformWindow will poll geometry() during construction below. Set the
+    // screen here so that high-dpi scaling will use the correct scale factor.
+    if (q->isTopLevel()) {
+        if (QScreen *screen = screenForGeometry(geometry))
+            setTopLevelScreen(screen, false);
+    }
+
     QPlatformIntegration *platformIntegration = QGuiApplicationPrivate::platformIntegration();
     platformWindow = nativeHandle ? platformIntegration->createForeignWindow(q, nativeHandle)
         : platformIntegration->createPlatformWindow(q);
@@ -1746,8 +1760,11 @@ void QWindow::setGeometry(const QRect &rect)
 
     d->positionPolicy = QWindowPrivate::WindowFrameExclusive;
     if (d->platformWindow) {
-        QRect nativeRect;
         QScreen *newScreen = d->screenForGeometry(rect);
+        if (newScreen && isTopLevel())
+            d->setTopLevelScreen(newScreen, true);
+
+        QRect nativeRect;
         if (newScreen && isTopLevel())
             nativeRect = QHighDpi::toNativePixels(rect, newScreen);
         else
diff --git a/src/gui/kernel/qwindowsysteminterface.cpp b/src/gui/kernel/qwindowsysteminterface.cpp
index 600cd9a7a52..4ab368d98cf 100644
--- a/src/gui/kernel/qwindowsysteminterface.cpp
+++ b/src/gui/kernel/qwindowsysteminterface.cpp
@@ -42,6 +42,7 @@
 #include "private/qguiapplication_p.h"
 #include "private/qevent_p.h"
 #include "private/qtouchdevice_p.h"
+#include "private/qscreen_p.h"
 #include <QAbstractEventDispatcher>
 #include <qpa/qplatformintegration.h>
 #include <qdebug.h>
@@ -822,6 +823,8 @@ void QWindowSystemInterface::handleScreenAdded(QPlatformScreen *ps, bool isPrima
         QGuiApplicationPrivate::screen_list.append(screen);
 
     QGuiApplicationPrivate::resetCachedDevicePixelRatio();
+    QHighDpiScaling::updateHighDpiScaling();
+    screen->d_func()->updateHighDpi();
 
     emit qGuiApp->screenAdded(screen);
 
@@ -879,6 +882,7 @@ void QWindowSystemInterface::handleScreenGeometryChange(QScreen *screen, const Q
 
 void QWindowSystemInterface::handleScreenLogicalDotsPerInchChange(QScreen *screen, qreal dpiX, qreal dpiY)
 {
+    QHighDpiScaling::updateHighDpiScaling();
     const QDpi effectiveDpi = QPlatformScreen::overrideDpi(QDpi{dpiX, dpiY});
     auto e = new QWindowSystemInterfacePrivate::ScreenLogicalDotsPerInchEvent(screen, effectiveDpi.first, effectiveDpi.second);
     QWindowSystemInterfacePrivate::handleWindowSystemEvent(e);
diff --git a/src/gui/painting/qbackingstore.cpp b/src/gui/painting/qbackingstore.cpp
index ebce163ed32..6143c5058de 100644
--- a/src/gui/painting/qbackingstore.cpp
+++ b/src/gui/painting/qbackingstore.cpp
@@ -66,6 +66,7 @@ public:
     QScopedPointer<QImage> highDpiBackingstore;
     QRegion staticContents;
     QSize size;
+    QSize nativeSize;
 };
 
 /*!
@@ -130,8 +131,7 @@ QWindow* QBackingStore::window() const
 
 void QBackingStore::beginPaint(const QRegion &region)
 {
-    if (d_ptr->highDpiBackingstore &&
-        d_ptr->highDpiBackingstore->devicePixelRatio() != d_ptr->window->devicePixelRatio())
+    if (d_ptr->nativeSize != QHighDpi::toNativePixels(size(), d_ptr->window))
         resize(size());
 
     QPlatformBackingStore *platformBackingStore = handle();
@@ -147,20 +147,23 @@ void QBackingStore::beginPaint(const QRegion &region)
     if (QHighDpiScaling::isActive() && device->devType() == QInternal::Image) {
         QImage *source = static_cast<QImage *>(device);
         const bool needsNewImage = d_ptr->highDpiBackingstore.isNull()
-            || source->data_ptr() != d_ptr->highDpiBackingstore->data_ptr()
+            || source->constBits() != d_ptr->highDpiBackingstore->constBits()
             || source->size() != d_ptr->highDpiBackingstore->size()
-            || source->devicePixelRatio() != d_ptr->highDpiBackingstore->devicePixelRatio();
+            || source->bytesPerLine() != d_ptr->highDpiBackingstore->bytesPerLine()
+            || source->format() != d_ptr->highDpiBackingstore->format();
         if (needsNewImage) {
             qCDebug(lcScaling) << "QBackingStore::beginPaint new backingstore for" << d_ptr->window;
             qCDebug(lcScaling) << "  source size" << source->size() << "dpr" << source->devicePixelRatio();
             d_ptr->highDpiBackingstore.reset(
                 new QImage(source->bits(), source->width(), source->height(), source->bytesPerLine(), source->format()));
-
-            qreal targetDevicePixelRatio = d_ptr->window->devicePixelRatio();
-            d_ptr->highDpiBackingstore->setDevicePixelRatio(targetDevicePixelRatio);
-            qCDebug(lcScaling) <<"  destination size" << d_ptr->highDpiBackingstore->size()
-                               << "dpr" << targetDevicePixelRatio;
         }
+
+        qreal targetDevicePixelRatio = d_ptr->window->devicePixelRatio();
+        d_ptr->highDpiBackingstore->setDevicePixelRatio(targetDevicePixelRatio);
+        qCDebug(lcScaling) <<"  destination size" << d_ptr->highDpiBackingstore->size()
+                           << "dpr" << targetDevicePixelRatio;
+    } else {
+        d_ptr->highDpiBackingstore.reset();
     }
 }
 
@@ -271,7 +274,8 @@ void QBackingStore::flush(const QRegion &region, QWindow *window, const QPoint &
 void QBackingStore::resize(const QSize &size)
 {
     d_ptr->size = size;
-    handle()->resize(QHighDpi::toNativePixels(size, d_ptr->window), d_ptr->staticContents);
+    d_ptr->nativeSize = QHighDpi::toNativePixels(size, d_ptr->window);
+    handle()->resize(d_ptr->nativeSize, d_ptr->staticContents);
 }
 
 /*!
diff --git a/src/plugins/platforms/windows/qwindowscontext.cpp b/src/plugins/platforms/windows/qwindowscontext.cpp
index 3e0059ca900..d9fb96d0aa6 100644
--- a/src/plugins/platforms/windows/qwindowscontext.cpp
+++ b/src/plugins/platforms/windows/qwindowscontext.cpp
@@ -218,6 +218,7 @@ void QWindowsUser32DLL::init()
         getWindowDpiAwarenessContext = (GetWindowDpiAwarenessContext)library.resolve("GetWindowDpiAwarenessContext");
         getAwarenessFromDpiAwarenessContext = (GetAwarenessFromDpiAwarenessContext)library.resolve("GetAwarenessFromDpiAwarenessContext");
         systemParametersInfoForDpi = (SystemParametersInfoForDpi)library.resolve("SystemParametersInfoForDpi");
+        getDpiForWindow = (GetDpiForWindow)library.resolve("GetDpiForWindow");
     }
 }
 
@@ -1443,12 +1444,16 @@ bool QWindowsContext::windowsProc(HWND hwnd, UINT message,
 #endif
     }   break;
     case QtWindows::DpiChangedEvent: {
+
+        const UINT dpi = HIWORD(wParam);
+        platformWindow->setSavedDpi(dpi);
+
         // Try to apply the suggested size first and then notify ScreenChanged
         // so that the resize event sent from QGuiApplication incorporates it
         // WM_DPICHANGED is sent with a size that avoids resize loops (by
         // snapping back to the previous screen, see QTBUG-65580).
         const bool doResize = resizeOnDpiChanged(platformWindow->window());
-        if (doResize) {
+        if (doResize && !platformWindow->testFlag(QWindowsWindow::WithinSetGeometry)) {
             platformWindow->setFlag(QWindowsWindow::WithinDpiChanged);
             platformWindow->updateFullFrameMargins();
             const auto prcNewWindow = reinterpret_cast<RECT *>(lParam);
@@ -1457,6 +1462,7 @@ bool QWindowsContext::windowsProc(HWND hwnd, UINT message,
             SetWindowPos(hwnd, nullptr, prcNewWindow->left, prcNewWindow->top,
                          prcNewWindow->right - prcNewWindow->left,
                          prcNewWindow->bottom - prcNewWindow->top, SWP_NOZORDER | SWP_NOACTIVATE);
+            platformWindow->handleGeometryChange();
             platformWindow->clearFlag(QWindowsWindow::WithinDpiChanged);
         }
         platformWindow->checkForScreenChanged(QWindowsWindow::FromDpiChange);
diff --git a/src/plugins/platforms/windows/qwindowscontext.h b/src/plugins/platforms/windows/qwindowscontext.h
index 8eca31e8bc5..2b63edd950c 100644
--- a/src/plugins/platforms/windows/qwindowscontext.h
+++ b/src/plugins/platforms/windows/qwindowscontext.h
@@ -107,6 +107,7 @@ struct QWindowsUser32DLL
     typedef int  (WINAPI *GetWindowDpiAwarenessContext)(HWND);
     typedef int  (WINAPI *GetAwarenessFromDpiAwarenessContext)(int);
     typedef BOOL (WINAPI *SystemParametersInfoForDpi)(UINT, UINT, PVOID, UINT, UINT);
+    typedef int  (WINAPI *GetDpiForWindow)(HWND);
 
     // Windows pointer functions (Windows 8 or later).
     EnableMouseInPointer enableMouseInPointer = nullptr;
@@ -123,6 +124,9 @@ struct QWindowsUser32DLL
     // Windows Vista onwards
     SetProcessDPIAware setProcessDPIAware = nullptr;
 
+    // Windows 10 version 1607 onwards
+    GetDpiForWindow getDpiForWindow = nullptr;
+
     // Clipboard listeners are present on Windows Vista onwards
     // but missing in MinGW 4.9 stub libs. Can be removed in MinGW 5.
     AddClipboardFormatListener addClipboardFormatListener = nullptr;
diff --git a/src/plugins/platforms/windows/qwindowsscreen.cpp b/src/plugins/platforms/windows/qwindowsscreen.cpp
index 6aeecdb7fa7..a1526ab7004 100644
--- a/src/plugins/platforms/windows/qwindowsscreen.cpp
+++ b/src/plugins/platforms/windows/qwindowsscreen.cpp
@@ -504,6 +504,28 @@ static void moveToVirtualScreen(QWindow *w, const QScreen *newScreen)
     w->setGeometry(geometry);
 }
 
+void QWindowsScreenManager::addScreen(const QWindowsScreenData &screenData)
+{
+    auto *newScreen = new QWindowsScreen(screenData);
+    m_screens.push_back(newScreen);
+    QWindowSystemInterface::handleScreenAdded(newScreen,
+                                              screenData.flags & QWindowsScreenData::PrimaryScreen);
+    qCDebug(lcQpaWindows) << "New Monitor: " << screenData;
+
+    // When a new screen is attached Window might move windows to the new screen
+    // automatically, in which case they will get a WM_DPICHANGED event. But at
+    // that point we have not received WM_DISPLAYCHANGE yet, so we fail to reflect
+    // the new screen's DPI. To account for this we explicitly check for screen
+    // change here, now that we are processing the WM_DISPLAYCHANGE.
+    const auto allWindows = QGuiApplication::allWindows();
+    for (QWindow *w : allWindows) {
+        if (w->isVisible() && w->handle() && w->type() != Qt::Desktop) {
+            if (QWindowsWindow *window = QWindowsWindow::windowsWindowOf(w))
+                window->checkForScreenChanged(QWindowsWindow::ScreenChangeMode::FromScreenAdded);
+        }
+    }
+}
+
 void QWindowsScreenManager::removeScreen(int index)
 {
     qCDebug(lcQpaWindows) << "Removing Monitor:" << m_screens.at(index)->data();
@@ -525,7 +547,7 @@ void QWindowsScreenManager::removeScreen(int index)
                     && (QWindowsWindow::baseWindowOf(w)->exStyle() & WS_EX_TOOLWINDOW)) {
                     moveToVirtualScreen(w, primaryScreen);
                 } else {
-                    QWindowSystemInterface::handleWindowScreenChanged(w, primaryScreen);
+                    QWindowSystemInterface::handleWindowScreenChanged<QWindowSystemInterface::SynchronousDelivery>(w, primaryScreen);
                 }
                 ++movedWindowCount;
             }
@@ -554,11 +576,7 @@ bool QWindowsScreenManager::handleScreenChanges()
             if (existingIndex == 0)
                 primaryScreenChanged = true;
         } else {
-            auto *newScreen = new QWindowsScreen(newData);
-            m_screens.push_back(newScreen);
-            QWindowSystemInterface::handleScreenAdded(newScreen,
-                                                             newData.flags & QWindowsScreenData::PrimaryScreen);
-            qCDebug(lcQpaWindows) << "New Monitor: " << newData;
+            addScreen(newData);
         }    // exists
     }        // for new screens.
     // Remove deleted ones but keep main monitors if we get only the
diff --git a/src/plugins/platforms/windows/qwindowsscreen.h b/src/plugins/platforms/windows/qwindowsscreen.h
index 5c095808f2d..917fce53b77 100644
--- a/src/plugins/platforms/windows/qwindowsscreen.h
+++ b/src/plugins/platforms/windows/qwindowsscreen.h
@@ -143,6 +143,7 @@ public:
     static bool isSingleScreen();
 
 private:
+    void addScreen(const QWindowsScreenData &screenData);
     void removeScreen(int index);
 
     WindowsScreenList m_screens;
diff --git a/src/plugins/platforms/windows/qwindowswindow.cpp b/src/plugins/platforms/windows/qwindowswindow.cpp
index 9d9180e4f57..769170ddadb 100644
--- a/src/plugins/platforms/windows/qwindowswindow.cpp
+++ b/src/plugins/platforms/windows/qwindowswindow.cpp
@@ -1389,6 +1389,8 @@ void QWindowsWindow::initialize()
         if (obtainedScreen && screen() != obtainedScreen)
             QWindowSystemInterface::handleWindowScreenChanged<QWindowSystemInterface::SynchronousDelivery>(w, obtainedScreen->screen());
     }
+    if (QWindowsContext::user32dll.getDpiForWindow)
+       setSavedDpi(QWindowsContext::user32dll.getDpiForWindow(handle()));
 }
 
 QSurfaceFormat QWindowsWindow::format() const
@@ -1949,10 +1951,11 @@ void QWindowsWindow::checkForScreenChanged(ScreenChangeMode mode)
         return;
     // For screens with different DPI: postpone until WM_DPICHANGE
     // Check on currentScreen as it can be 0 when resuming a session (QTBUG-80436).
-    if (mode == FromGeometryChange && currentScreen != nullptr
-        && !equalDpi(currentScreen->logicalDpi(), newScreen->logicalDpi())) {
+    const QDpi currentDpi = savedDpi() ? QDpi(savedDpi(), savedDpi()) : currentScreen ? currentScreen->logicalDpi() : 0;
+    const bool changingDpi = !equalDpi(currentDpi, newScreen->logicalDpi());
+    if (mode == FromGeometryChange && currentScreen != nullptr && changingDpi)
         return;
-    }
+
     qCDebug(lcQpaWindows).noquote().nospace() << __FUNCTION__
         << ' ' << window() << " \"" << (currentScreen ? currentScreen->name() : QString())
         << "\"->\"" << newScreen->name() << '"';
diff --git a/src/plugins/platforms/windows/qwindowswindow.h b/src/plugins/platforms/windows/qwindowswindow.h
index ac207aa48f4..5401597c66f 100644
--- a/src/plugins/platforms/windows/qwindowswindow.h
+++ b/src/plugins/platforms/windows/qwindowswindow.h
@@ -264,6 +264,7 @@ public:
     void propagateSizeHints() override;
     static bool handleGeometryChangingMessage(MSG *message, const QWindow *qWindow, const QMargins &marginsDp);
     bool handleGeometryChanging(MSG *message) const;
+    void handleGeometryChange();
     QMargins frameMargins() const override;
     QMargins fullFrameMargins() const override;
     void setFullFrameMargins(const QMargins &newMargins);
@@ -346,7 +347,7 @@ public:
     void alertWindow(int durationMs = 0);
     void stopAlertWindow();
 
-    enum ScreenChangeMode { FromGeometryChange, FromDpiChange };
+    enum ScreenChangeMode { FromGeometryChange, FromDpiChange, FromScreenAdded };
     void checkForScreenChanged(ScreenChangeMode mode = FromGeometryChange);
 
     static void setTouchWindowTouchTypeStatic(QWindow *window, QWindowsWindowFunctions::TouchWindowTouchTypes touchTypes);
@@ -359,6 +360,9 @@ public:
     static const char *embeddedNativeParentHandleProperty;
     static const char *hasBorderInFullScreenProperty;
 
+    void setSavedDpi(int dpi) { m_savedDpi = dpi; }
+    int savedDpi() const { return m_savedDpi; }
+
 private:
     inline void show_sys() const;
     inline QWindowsWindowData setWindowFlags_sys(Qt::WindowFlags wt, unsigned flags = 0) const;
@@ -370,7 +374,6 @@ private:
     inline bool isDropSiteEnabled() const { return m_dropTarget != nullptr; }
     void setDropSiteEnabled(bool enabled);
     void updateDropSite(bool topLevel);
-    void handleGeometryChange();
     void handleWindowStateChange(Qt::WindowStates state);
     inline void destroyIcon();
     void fireExpose(const QRegion &region, bool force=false);
@@ -392,6 +395,7 @@ private:
     HICON m_iconSmall = nullptr;
     HICON m_iconBig = nullptr;
     void *m_surface = nullptr;
+    int m_savedDpi = 0;
 
     static bool m_screenForGLInitialized;
 
