diff --git a/td/generate/scheme/td_api.tl b/td/generate/scheme/td_api.tl
index 964df35a..95b1e135 100644
--- a/td/generate/scheme/td_api.tl
+++ b/td/generate/scheme/td_api.tl
@@ -4584,7 +4584,7 @@ getChatFilterDefaultIconName filter:chatFilter = Text;
 setChatTitle chat_id:int53 title:string = Ok;
 
 //@description Changes the photo of a chat. Supported only for basic groups, supergroups and channels. Requires can_change_info administrator right
-//@chat_id Chat identifier @photo New chat photo. Pass null to delete the chat photo
+//@chat_id Chat identifier @photo New chat photo. Pass null to delete the chat photo; may be null.
 setChatPhoto chat_id:int53 photo:InputChatPhoto = Ok;
 
 //@description Changes the message TTL setting (sets a new self-destruct timer) in a chat. Requires can_delete_messages administrator right in basic groups, supergroups and channels
