diff --git a/src/gui/kernel/qhighdpiscaling.cpp b/src/gui/kernel/qhighdpiscaling.cpp
index a0e1b48dcb..f01b989711 100644
--- a/src/gui/kernel/qhighdpiscaling.cpp
+++ b/src/gui/kernel/qhighdpiscaling.cpp
@@ -630,7 +630,7 @@ qreal QHighDpiScaling::screenSubfactor(const QPlatformScreen *screen)
     if (!screenPropertyUsed && m_usePlatformPluginDpi)
         factor = roundScaleFactor(rawScaleFactor(screen));
 
-    return factor;
+    return factor * (roundScaleFactor(screen->devicePixelRatio()) / screen->devicePixelRatio());
 }
 
 QDpi QHighDpiScaling::logicalDpi(const QScreen *screen)
@@ -647,7 +647,9 @@ QDpi QHighDpiScaling::logicalDpi(const QScreen *screen)
 
     const qreal scaleFactor = rawScaleFactor(screen->handle());
     const qreal roundedScaleFactor = roundScaleFactor(scaleFactor);
-    return effectiveLogicalDpi(screen->handle(), scaleFactor, roundedScaleFactor);
+    const qreal devicePixelRatio = screen->handle()->devicePixelRatio();
+    const qreal roundedDevicePixelRatio = roundScaleFactor(devicePixelRatio);
+    return effectiveLogicalDpi(screen->handle(), scaleFactor * devicePixelRatio, roundedScaleFactor * roundedDevicePixelRatio);
 }
 
 // Returns the screen containing \a position, using \a guess as a starting point
