diff --git a/src/xml/dom/qdom.cpp b/src/xml/dom/qdom.cpp
index 85718cd5e0..5ef55b4f93 100644
--- a/src/xml/dom/qdom.cpp
+++ b/src/xml/dom/qdom.cpp
@@ -5914,7 +5914,7 @@ void QDomDocumentPrivate::saveDocument(QTextStream& s, const int indent, QDomNod
                 if (!encoding)
                     qWarning() << "QDomDocument::save(): Unsupported encoding" << enc << "specified.";
                 else
-                    s.setEncoding(encoding.value());
+                    s.setEncoding(*encoding);
             }
         }
 #endif
