diff --git a/src/gui/platform/unix/qgenericunixservices.cpp b/src/gui/platform/unix/qgenericunixservices.cpp
index 34d9c37e24..a1cf09aa53 100644
--- a/src/gui/platform/unix/qgenericunixservices.cpp
+++ b/src/gui/platform/unix/qgenericunixservices.cpp
@@ -162,7 +162,7 @@ static inline bool launch(const QString &launcher, const QUrl &url,
 #if QT_CONFIG(dbus)
 static inline bool checkNeedPortalSupport()
 {
-    return QFileInfo::exists("/.flatpak-info"_L1) || qEnvironmentVariableIsSet("SNAP");
+    return true;
 }
 
 static inline QDBusMessage xdgDesktopPortalOpenFile(const QUrl &url, const QString &parentWindow,
