diff --git a/src/client/qwaylandinputdevice.cpp b/src/client/qwaylandinputdevice.cpp
index 57c0ee8a..4ae62eb4 100644
--- a/src/client/qwaylandinputdevice.cpp
+++ b/src/client/qwaylandinputdevice.cpp
@@ -215,7 +215,7 @@ public:
     {
         int scale = 0;
         for (auto *screen : m_screens)
-            scale = qMax(scale, screen->scale());
+            scale = qMax(scale, qCeil(screen->scale()));
         return scale;
     }
 
diff --git a/src/client/qwaylandscreen.cpp b/src/client/qwaylandscreen.cpp
index 4c9e46e0..b2063043 100644
--- a/src/client/qwaylandscreen.cpp
+++ b/src/client/qwaylandscreen.cpp
@@ -178,14 +178,16 @@ Qt::ScreenOrientation QWaylandScreen::orientation() const
     return m_orientation;
 }
 
-int QWaylandScreen::scale() const
+qreal QWaylandScreen::scale() const
 {
-    return mScale;
+    return devicePixelRatio();
 }
 
 qreal QWaylandScreen::devicePixelRatio() const
 {
-    return qreal(mScale);
+    return (mWaylandDisplay->fractionalScaleManager() && std::fmod(mFractionalScale, qreal(1)) != 0)
+        ? mFractionalScale
+        : qreal(mScale);
 }
 
 qreal QWaylandScreen::refreshRate() const
@@ -370,7 +372,12 @@ void QWaylandScreen::zxdg_output_v1_name(const QString &name)
 void QWaylandScreen::updateXdgOutputProperties()
 {
     Q_ASSERT(zxdg_output_v1::isInitialized());
+    mFractionalScale = std::round((qreal(mGeometry.width()) / qreal(mXdgGeometry.width())
+        + qreal(mGeometry.height()) / qreal(mXdgGeometry.height()))
+        * qreal(0.5) * qreal(100)) / qreal(100);
+    const QDpi dpi = logicalDpi();
     QWindowSystemInterface::handleScreenGeometryChange(screen(), geometry(), geometry());
+    QWindowSystemInterface::handleScreenLogicalDotsPerInchChange(screen(), dpi.first, dpi.second);
 }
 
 } // namespace QtWaylandClient
diff --git a/src/client/qwaylandscreen_p.h b/src/client/qwaylandscreen_p.h
index 8cc9d05d..3b4e4780 100644
--- a/src/client/qwaylandscreen_p.h
+++ b/src/client/qwaylandscreen_p.h
@@ -65,7 +65,7 @@ public:
     QWindow *topLevelAt(const QPoint &point) const override;
 
     Qt::ScreenOrientation orientation() const override;
-    int scale() const;
+    qreal scale() const;
     qreal devicePixelRatio() const override;
     qreal refreshRate() const override;
 
@@ -122,6 +122,7 @@ protected:
     QRect mGeometry;
     QRect mXdgGeometry;
     int mScale = 1;
+    qreal mFractionalScale = 1;
     int mDepth = 32;
     int mRefreshRate = 60000;
     int mSubpixel = -1;
diff --git a/src/client/qwaylandwindow.cpp b/src/client/qwaylandwindow.cpp
index 42fb1312..26c611dd 100644
--- a/src/client/qwaylandwindow.cpp
+++ b/src/client/qwaylandwindow.cpp
@@ -57,7 +57,7 @@ QWaylandWindow::QWaylandWindow(QWindow *window, QWaylandDisplay *display)
             mFrameCallbackTimeout = frameCallbackTimeout;
     }
 
-    mScale = waylandScreen() ? waylandScreen()->scale() : 1; // fallback to 1 if we don't have a real screen
+    mScale = waylandScreen() ? waylandScreen()->scale() : qreal(1); // fallback to 1 if we don't have a real screen
 
     static WId id = 1;
     mWindowId = id++;
@@ -103,7 +103,7 @@ void QWaylandWindow::initWindow()
         initializeWlSurface();
     }
 
-    if (mDisplay->fractionalScaleManager() && qApp->highDpiScaleFactorRoundingPolicy() == Qt::HighDpiScaleFactorRoundingPolicy::PassThrough) {
+    if (mDisplay->fractionalScaleManager()) {
         mFractionalScale.reset(new QWaylandFractionalScale(mDisplay->fractionalScaleManager()->get_fractional_scale(mSurface->object())));
 
         connect(mFractionalScale.data(), &QWaylandFractionalScale::preferredScaleChanged,
@@ -1438,7 +1438,7 @@ void QWaylandWindow::updateScale()
         return;
     }
 
-    int scale = mLastReportedScreen->isPlaceholder() ? 1 : static_cast<QWaylandScreen *>(mLastReportedScreen)->scale();
+    qreal scale = mLastReportedScreen->isPlaceholder() ? qreal(1) : static_cast<QWaylandScreen *>(mLastReportedScreen)->scale();
     setScale(scale);
 }
 
@@ -1467,7 +1467,7 @@ void QWaylandWindow::setScale(qreal newScale)
 #if QT_CONFIG(cursor)
 void QWaylandWindow::setMouseCursor(QWaylandInputDevice *device, const QCursor &cursor)
 {
-    int fallbackBufferScale = int(devicePixelRatio());
+    int fallbackBufferScale = qCeil(devicePixelRatio());
     device->setCursor(&cursor, {}, fallbackBufferScale);
 }
 
