diff --git a/src/plugins/platforms/cocoa/qcocoaapplicationdelegate.mm b/src/plugins/platforms/cocoa/qcocoaapplicationdelegate.mm
index e8d789275c..9981588219 100644
--- a/src/plugins/platforms/cocoa/qcocoaapplicationdelegate.mm
+++ b/src/plugins/platforms/cocoa/qcocoaapplicationdelegate.mm
@@ -138,7 +138,12 @@ QT_USE_NAMESPACE
 
 - (NSMenu *)applicationDockMenu:(NSApplication *)sender
 {
-    Q_UNUSED(sender);
+    // Patch: We need to catch that event in delegate.
+    if (reflectionDelegate
+        && [reflectionDelegate respondsToSelector:@selector(applicationDockMenu:)]) {
+        return [reflectionDelegate applicationDockMenu:sender];
+    }
+
     // Manually invoke the delegate's -menuWillOpen: method.
     // See QTBUG-39604 (and its fix) for details.
     [self.dockMenu.delegate menuWillOpen:self.dockMenu];
