diff --git a/src/gui/text/qtextlayout.cpp b/src/gui/text/qtextlayout.cpp
index 8d105d9523..4e8ef32854 100644
--- a/src/gui/text/qtextlayout.cpp
+++ b/src/gui/text/qtextlayout.cpp
@@ -3105,8 +3105,19 @@ int QTextLine::xToCursor(qreal _x, CursorPosition cpos) const
     bool visual = eng->visualCursorMovement();
     if (x <= 0) {
         // left of first item
+
+        // Patch: Fix go up one line in RTL case of an empty line.
+        // except for the last line we assume that the
+        // character between lines is a space and we want
+        // to position the cursor to the left of that
+        // character.
+        int maxPos = line.from + line_length;
+        if (index < eng->lines.count() - 1)
+            maxPos = qMin(eng->previousLogicalPosition(maxPos), maxPos);
+
         if (eng->isRightToLeft())
-            return line.from + line_length;
+            //return line.from + line_length;
+            return maxPos;
         return line.from;
     }   else if (x < line.textWidth || (line.justified && x < line.width)) {
         // has to be in one of the runs
