diff --git a/src/platformsupport/fontdatabases/freetype/qfontengine_ft.cpp b/src/platformsupport/fontdatabases/freetype/qfontengine_ft.cpp
index e132442e37..5ad038992c 100644
--- a/src/platformsupport/fontdatabases/freetype/qfontengine_ft.cpp
+++ b/src/platformsupport/fontdatabases/freetype/qfontengine_ft.cpp
@@ -148,7 +148,7 @@ QtFreetypeData *qt_getFreetypeData()
         freetypeData = new QtFreetypeData;
     if (!freetypeData->library) {
         FT_Init_FreeType(&freetypeData->library);
-#if defined(FT_FONT_FORMATS_H)
+#if 0
         // Freetype defaults to disabling stem-darkening on CFF, we re-enable it.
         FT_Bool no_darkening = false;
         FT_Property_Set(freetypeData->library, "cff", "no-stem-darkening", &no_darkening);
@@ -882,7 +882,7 @@ bool QFontEngineFT::init(FaceId faceId, bool antialias, GlyphFormat format,
             }
         }
     }
-#if defined(FT_FONT_FORMATS_H)
+#if 0
     const char *fmt = FT_Get_Font_Format(face);
     if (fmt && qstrncmp(fmt, "CFF", 4) == 0) {
         FT_Bool no_stem_darkening = true;
