diff --git a/src/plugins/hardwareintegration/client/wayland-egl/main.cpp b/src/plugins/hardwareintegration/client/wayland-egl/main.cpp
index 11bf5806..30ad2d62 100644
--- a/src/plugins/hardwareintegration/client/wayland-egl/main.cpp
+++ b/src/plugins/hardwareintegration/client/wayland-egl/main.cpp
@@ -40,6 +40,8 @@
 #include <QtWaylandClient/private/qwaylandclientbufferintegrationplugin_p.h>
 #include <QtWaylandEglClientHwIntegration/private/qwaylandeglclientbufferintegration_p.h>
 
+extern "C" bool epoxy_load_egl(bool exit_if_fails, bool load);
+
 QT_BEGIN_NAMESPACE
 
 namespace QtWaylandClient {
@@ -54,6 +56,10 @@ public:
 
 QWaylandClientBufferIntegration *QWaylandEglClientBufferPlugin::create(const QString&, const QStringList&)
 {
+    if (!epoxy_load_egl(false, true)) {
+        return nullptr;
+    }
+
     return new QWaylandEglClientBufferIntegration();
 }
 
