diff --git a/src/gui/text/windows/qwindowsfontdatabasebase.cpp b/src/gui/text/windows/qwindowsfontdatabasebase.cpp
index 036c3c37582..18679ef42da 100644
--- a/src/gui/text/windows/qwindowsfontdatabasebase.cpp
+++ b/src/gui/text/windows/qwindowsfontdatabasebase.cpp
@@ -967,9 +967,13 @@ QString QWindowsFontDatabaseBase::familyForStyleHint(QFont::StyleHint styleHint)
 
 static const char *other_tryFonts[] = {
     "Arial",
+    "Yu Gothic UI",
     "MS UI Gothic",
+    "Malgun Gothic",
     "Gulim",
+    "Microsoft YaHei UI",
     "SimSun",
+    "Microsoft JhengHei UI",
     "PMingLiU",
     "Arial Unicode MS",
     0
@@ -1023,6 +1027,7 @@ QStringList QWindowsFontDatabaseBase::extraTryFontsForFamily(const QString &fami
 {
     QStringList result;
     if (!QFontDatabase::writingSystems(family).contains(QFontDatabase::Symbol)) {
+        result << systemDefaultFont().family();
         if (!tryFonts) {
             LANGID lid = GetUserDefaultLangID();
             switch (lid&0xff) {
