diff --git a/src/client/configure.cmake b/src/client/configure.cmake
index 11612d11..2c3d7727 100644
--- a/src/client/configure.cmake
+++ b/src/client/configure.cmake
@@ -184,7 +184,7 @@ qt_feature("wayland-client-xdg-shell" PRIVATE
 )
 qt_feature("egl-extension-platform-wayland" PRIVATE
     LABEL "EGL wayland platform extension"
-    CONDITION QT_FEATURE_wayland_client AND QT_FEATURE_opengl AND QT_FEATURE_egl AND TEST_egl_1_5_wayland
+    CONDITION 0
 )
 qt_feature("wayland-egl" PRIVATE
     LABEL "EGL"
