diff --git a/src/client/qwaylandshmbackingstore.cpp b/src/client/qwaylandshmbackingstore.cpp
index 02d830ad..aadebac0 100644
--- a/src/client/qwaylandshmbackingstore.cpp
+++ b/src/client/qwaylandshmbackingstore.cpp
@@ -10,6 +10,7 @@
 #include <QtCore/qdebug.h>
 #include <QtCore/qstandardpaths.h>
 #include <QtCore/qtemporaryfile.h>
+#include <QtGui/QBackingStore>
 #include <QtGui/QPainter>
 #include <QtGui/QTransform>
 #include <QMutexLocker>
@@ -171,7 +172,7 @@ void QWaylandShmBackingStore::beginPaint(const QRegion &region)
     mPainting = true;
     ensureSize();
 
-    if (mBackBuffer->image()->hasAlphaChannel()) {
+    if (backingStore()->paintDevice() == paintDevice() && mBackBuffer->image()->hasAlphaChannel()) {
         QPainter p(paintDevice());
         p.setCompositionMode(QPainter::CompositionMode_Source);
         const QColor blank = Qt::transparent;
