diff --git a/src/plugins/platforms/cocoa/qcocoadrag.mm b/src/plugins/platforms/cocoa/qcocoadrag.mm
index fa073e5e29..5cdb0454ff 100644
--- a/src/plugins/platforms/cocoa/qcocoadrag.mm
+++ b/src/plugins/platforms/cocoa/qcocoadrag.mm
@@ -130,7 +130,7 @@ Qt::DropAction QCocoaDrag::drag(QDrag *o)
     m_drag = o;
     m_executed_drop_action = Qt::IgnoreAction;
 
-    QMacPasteboard dragBoard(CFStringRef(NSPasteboardNameDrag), QMacInternalPasteboardMime::MIME_DND);
+    QMacPasteboard dragBoard(CFStringRef(NSDragPboard), QMacInternalPasteboardMime::MIME_DND);
     m_drag->mimeData()->setData(QLatin1String("application/x-qt-mime-type-name"), QByteArray("dummy"));
     dragBoard.setMimeData(m_drag->mimeData(), QMacPasteboard::LazyRequest);
 
@@ -149,7 +149,7 @@ Qt::DropAction QCocoaDrag::drag(QDrag *o)
     CGFloat flippedY = dragImage.size.height - hotSpot.y();
     event_location.y -= flippedY;
     NSSize mouseOffset_unused = NSMakeSize(0.0, 0.0);
-    NSPasteboard *pboard = [NSPasteboard pasteboardWithName:NSPasteboardNameDrag];
+    NSPasteboard *pboard = [NSPasteboard pasteboardWithName:NSDragPboard];
 
     [theWindow dragImage:dragImage
         at:event_location
@@ -182,7 +182,7 @@ bool QCocoaDrag::maybeDragMultipleItems()
     auto *sourceView = static_cast<NSView<NSDraggingSource>*>(view);
 
     const auto &qtUrls = m_drag->mimeData()->urls();
-    NSPasteboard *dragBoard = [NSPasteboard pasteboardWithName:NSPasteboardNameDrag];
+    NSPasteboard *dragBoard = [NSPasteboard pasteboardWithName:NSDragPboard];
 
     if (qtUrls.size() <= 1) {
         // Good old -dragImage: works perfectly for this ...
