diff --git a/src/gui/platform/unix/qgnometheme.cpp b/src/gui/platform/unix/qgnometheme.cpp
index a9fb85e15bc..b7d246a8947 100644
--- a/src/gui/platform/unix/qgnometheme.cpp
+++ b/src/gui/platform/unix/qgnometheme.cpp
@@ -105,7 +105,7 @@ QVariant QGnomeTheme::themeHint(QPlatformTheme::ThemeHint hint) const
     case QPlatformTheme::SystemIconThemeName:
         return QVariant(QStringLiteral("Adwaita"));
     case QPlatformTheme::SystemIconFallbackThemeName:
-        return QVariant(QStringLiteral("gnome"));
+        return QVariant(QStringLiteral("hicolor"));
     case QPlatformTheme::IconThemeSearchPaths:
         return QVariant(xdgIconThemePaths());
     case QPlatformTheme::IconPixmapSizes:
diff --git a/src/plugins/platformthemes/gtk3/qgtk3theme.cpp b/src/plugins/platformthemes/gtk3/qgtk3theme.cpp
index c720fc41dbd..7055a18622c 100644
--- a/src/plugins/platformthemes/gtk3/qgtk3theme.cpp
+++ b/src/plugins/platformthemes/gtk3/qgtk3theme.cpp
@@ -96,7 +96,6 @@ QGtk3Theme::QGtk3Theme()
     SETTING_CONNECT("gtk-entry-password-hint-timeout");
     SETTING_CONNECT("gtk-dnd-drag-threshold");
     SETTING_CONNECT("gtk-icon-theme-name");
-    SETTING_CONNECT("gtk-fallback-icon-theme");
     SETTING_CONNECT("gtk-font-name");
     SETTING_CONNECT("gtk-application-prefer-dark-theme");
     SETTING_CONNECT("gtk-theme-name");
@@ -140,8 +139,6 @@ QVariant QGtk3Theme::themeHint(QPlatformTheme::ThemeHint hint) const
         return QVariant(gtkSetting<gint>("gtk-dnd-drag-threshold"));
     case QPlatformTheme::SystemIconThemeName:
         return QVariant(gtkSetting("gtk-icon-theme-name"));
-    case QPlatformTheme::SystemIconFallbackThemeName:
-        return QVariant(gtkSetting("gtk-fallback-icon-theme"));
     case QPlatformTheme::MouseCursorTheme:
         return QVariant(gtkSetting("gtk-cursor-theme-name"));
     case QPlatformTheme::MouseCursorSize: {
