diff --git a/src/gui/CMakeLists.txt b/src/gui/CMakeLists.txt
index 8c0f11924d1..4cbca9a7dea 100644
--- a/src/gui/CMakeLists.txt
+++ b/src/gui/CMakeLists.txt
@@ -58,7 +58,7 @@ if(QT_FEATURE_opengl)
 endif()
 
 qt_internal_add_module(Gui
-    PLUGIN_TYPES accessiblebridge platforms platforms/darwin xcbglintegrations platformthemes platforminputcontexts generic iconengines imageformats egldeviceintegrations
+    PLUGIN_TYPES accessiblebridge platforms platforms/darwin xcbglintegrations platformthemes platforminputcontexts generic iconengines imageformats egldeviceintegrations wayland-decoration-client wayland-graphics-integration-client wayland-shell-integration
     QMAKE_MODULE_CONFIG "${qmake_module_config}"
     SOURCES
         compat/removed_api.cpp
